{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "21c9b44b-9032-488f-a68c-ef0af006f7fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import logging\n",
    "from ImmuneGWAS.helpers.getpaths import get_paths, get_sumstats_path\n",
    "from ImmuneGWAS.helpers import dbsnp, ldlink\n",
    "import ImmuneGWAS.resources.immune_GWAS as immune_GWAS\n",
    "from ImmuneGWAS import config\n",
    "from ImmuneGWAS.variant import Variant\n",
    "\n",
    "from ImmuneGWAS.resources.eqtlgen import eqtlgen_trans_LDblock_query \n",
    "from ImmuneGWAS.resources.eqtlgen import eqtlgen_cis_LDblock_query \n",
    "from ImmuneGWAS.resources.tokyo_eqtl import tokyo_eqtl_LDblock_query\n",
    "from ImmuneGWAS.resources.eqtl_cat import *\n",
    "from IPython.display import display_html \n",
    "from IPython.display import display,HTML\n",
    "\n",
    "from scipy.cluster import hierarchy\n",
    "\n",
    "from ImmuneGWAS.helpers import get_gene_symbol  # convert ensembl to gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7e42f553-ec06-4d45-89ee-4132a647f95d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ldpruned_hits = \"/media/cbio3/projects/antton/Immune_cell_GWAS/data/hits_only_table_hg38_LDprunned.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f00a739-068a-4767-a88f-721b38c7cbea",
   "metadata": {},
   "source": [
    "Designing the lookup:\n",
    "\n",
    "1) Variant object calls trans_df lookup\n",
    "2) Harmonize according to EA/OA, raise exceptions for multi-allelic SNPs\n",
    "3) Split trans_df into up and down\n",
    "4) Return row-scaled heatmap df where cell ordering is hard-coded and rows are resorted for clustering\n",
    "5) Plotting function for heatmap df\n",
    "6) Get flow phenotypes\n",
    "7) Get LDTrait phenotypes\n",
    "8) Summarize phenotpyes into tables with betas (remove duplicate phenotypes etc)\n",
    "9) Get a single summary cis-eQTL table\n",
    "10) Return summary report - 2 heatmaps for up and down, cis-eQTLs, phenotypes summarized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d355c9bd-7599-46f1-9f16-dff7f9ab3a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid = \"rs1354034\"\n",
    "chrom = 3\n",
    "pos = 56815721\n",
    "EA = 'T'\n",
    "OA = 'C'\n",
    "\n",
    "x = Variant(rsid, chrom, pos, EA, OA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d1f08a26-1316-4e3f-b48b-95319325b017",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.get_gwas_phenotypes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9cda26a3-fab4-42c9-91d1-84aacebf4c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def multi_column_df_display(list_dfs, cols=3):\n",
    "    html_table = \"<table style='width:100%; border:0px'>{content}</table>\"\n",
    "    html_row = \"<tr style='border:0px'>{content}</tr>\"\n",
    "    html_cell = \"<td style='width:{width}%;vertical-align:top;border:0px'>{{content}}</td>\"\n",
    "    html_cell = html_cell.format(width=100/cols)\n",
    "\n",
    "    cells = [ html_cell.format(content=df.to_html()) for df in list_dfs ]\n",
    "    cells += (cols - (len(list_dfs)%cols)) * [html_cell.format(content=\"\")] # pad\n",
    "    rows = [ html_row.format(content=\"\".join(cells[i:i+cols])) for i in range(0,len(cells),cols)]\n",
    "    display(HTML(html_table.format(content=\"\".join(rows))))\n",
    "\n",
    "def summary_trans(var: Variant):\n",
    "    \n",
    "    # retrieve trans-eQTLs\n",
    "    trans = eqtlgen_trans_LDblock_query(var)\n",
    "    trans.Zscore = trans.Zscore.astype(float)\n",
    "    \n",
    "    \n",
    "    # check EA for sign of beta\n",
    "    \n",
    "    if set(trans.AssessedAllele == x.EA) == {True}:\n",
    "        trans.Zscore = trans.Zscore * -1\n",
    "    \n",
    "    # split into up and down genes\n",
    "    up = trans[trans.Zscore>0]\n",
    "    down = trans[trans.Zscore<0]\n",
    "    up = up.Gene\n",
    "    down = down.Gene\n",
    "    \n",
    "    # dictionary to map ensembl to gene name\n",
    "    mapper = dict(zip(trans.Gene, trans.GeneSymbol))\n",
    "    \n",
    "    #get ge matrix from ImmunexUT\n",
    "    paths = get_paths(root=config.cbio_root)\n",
    "    ge = pd.read_csv(paths['ge_tokyo'], sep='\\t')\n",
    "    ge = ge.set_index(\"Gene_id\")\n",
    "    ge = ge.drop(\"Gene_name\", axis=1)\n",
    "    ge = ge.drop_duplicates()\n",
    "    \n",
    "    # get indices for trans-eQTL matrix\n",
    "    # note any missing genes will be dropped\n",
    "    \n",
    "    u = [i for i in up if i in ge.index]\n",
    "    u = set(u)\n",
    "    d = [i for i in down if i in ge.index]\n",
    "    d = set(d)\n",
    "    \n",
    "    # filter gene expression matrix into up and down genes\n",
    "    u = ge.loc[u]\n",
    "    d = ge.loc[d]\n",
    "    \n",
    "    # rename ensembl IDs to gene names\n",
    "    u.index = u.index.map(mapper)\n",
    "    d.index = d.index.map(mapper)\n",
    "\n",
    "    # hard-code order of cell types according to gene expression similarity on whole GE matrix\n",
    "    \n",
    "    ordered_cols = ['Plasmablast', 'LDG', 'Neu', 'pDC', 'mDC', 'CL_Mono', 'Int_Mono',\n",
    "           'CD16p_Mono', 'NC_Mono', 'Naive_B', 'USM_B', 'DN_B', 'SM_B', 'NK',\n",
    "           'Mem_CD8', 'EM_CD8', 'TEMRA_CD8', 'Naive_CD8', 'Fr_I_nTreg',\n",
    "           'Naive_CD4', 'Fr_III_T', 'Fr_II_eTreg', 'CM_CD8', 'Th2', 'Tfh', 'Th17',\n",
    "           'Mem_CD4', 'Th1']\n",
    "    \n",
    "    # normalize up and down genes\n",
    "\n",
    "    def standard_scale(x):\n",
    "        return (x-x.min())/x.max()\n",
    "\n",
    "    u = u.apply(standard_scale, axis=1)\n",
    "    d = d.apply(standard_scale, axis=1)\n",
    "\n",
    "    # reorder columns\n",
    "    \n",
    "    u = u.loc[:, ordered_cols]\n",
    "    d = d.loc[:, ordered_cols]\n",
    "    \n",
    "    # reoder rows by dendrogram\n",
    "    if len(u)>1:\n",
    "        Z = hierarchy.linkage(u, optimal_ordering=True, method='ward')\n",
    "        dendro = hierarchy.dendrogram(Z, no_plot=True)\n",
    "    \n",
    "        u = u.iloc[dendro['leaves'],:]\n",
    "    \n",
    "    if len(d)>1:\n",
    "        Z = hierarchy.linkage(d, optimal_ordering=True, method='ward')\n",
    "        dendro = hierarchy.dendrogram(Z, no_plot=True)\n",
    "        d = d.iloc[dendro['leaves'],:]\n",
    "    # collect all my cis-eqtls\n",
    "\n",
    "    tokyo_eqtl_LDblock_query(x)\n",
    "    eqtlgen_cis_LDblock_query (x)\n",
    "    eqtl_catalogue_LDblock_query_type_restricted_multitype(x)\n",
    "\n",
    "    # now all of these are in the Results object\n",
    "\n",
    "    gwas = ldlink.ldtrait(x)\n",
    "\n",
    "    # first try to filter by the same SNP as in Variant object, otherwise filter down to one rsid\n",
    "    \n",
    "    tokyo = x.results.tokyo_eqtl()[x.results.tokyo_eqtl()['Variant_ID']==x.rsid]\n",
    "    \n",
    "    if len(tokyo)>0:\n",
    "        tokyo = tokyo[['Gene_name','Variant_ID', 'Backward_slope', 'Backward_P', 'cell_type']]\n",
    "        tokyo = tokyo.sort_values(by='Gene_name')\n",
    "    \n",
    "    elif len(tokyo)==0:\n",
    "        tokyo = x.results.tokyo_eqtl()\n",
    "        rsid = tokyo['Variant ID'][0]\n",
    "        \n",
    "\n",
    "    \n",
    "    cat = x.results.eqtl_cat()\n",
    "    cat = cat[cat.position==str(x.pos)]\n",
    "    cat = cat[['gene_symbol', 'z', 'study', 'cell_type']]\n",
    "    \n",
    "    \n",
    "    gwas_traits = set(x.results.ldtrait()['GWAS Trait'].tolist())\n",
    "    \n",
    "    \n",
    "    \n",
    "    cisgen = x.results.eqtlgen_cis()\n",
    "    cisgen = cisgen[cisgen.SNP==x.rsid]\n",
    "    if len(cisgen)>0:\n",
    "        cisgen = cisgen[['SNP', 'Zscore', 'Pvalue', 'GeneSymbol']].sort_values(by='GeneSymbol')\n",
    "    elif len(csigen)==0:\n",
    "        cisgen = x.results.eqtlgen_cis()\n",
    "        rsid = cisgen.SNP[0]\n",
    "        cisgen = cisgen[cisgen.SNP==rsid]\n",
    "        cisgen = cisgen[['SNP', 'Zscore', 'Pvalue', 'GeneSymbol']].sort_values(by='GeneSymbol')\n",
    "    # get the cis-eqtl genes to plot a heatmap of the cis-eQTL genes\n",
    "    \n",
    "    a = x.results.eqtlgen_cis_df['Gene']\n",
    "\n",
    "    b = x.results.tokyo_eqtl_df['Gene_id'].apply(lambda x:x.split(\".\")[0])\n",
    "\n",
    "    c = x.results.eqtl_cat_df['molecular_trait_id']\n",
    "    \n",
    "    cis_genes = list(set(pd.concat([a,b,c]).tolist()))\n",
    "    cis_genes = [i for i in cis_genes if i in ge.index]\n",
    "    cis_genes = ge.loc[cis_genes]\n",
    "    cis_genes = cis_genes.apply(standard_scale, axis=1)\n",
    "    cis_genes = cis_genes.loc[:, ordered_cols]\n",
    "    cis_genes.index = cis_genes.index.map(get_gene_symbol)\n",
    "    \n",
    "    list_dfs = [tokyo, cat, cisgen]\n",
    "\n",
    "    display(tokyo)\n",
    "    display(cat)\n",
    "    display(cisgen)\n",
    "    print(\"GWAS Catalog\\n\")\n",
    "    display(gwas_traits)\n",
    "    print(\"Flow phenos\\n\")\n",
    "    display(x.get_gwas_phenotypes())\n",
    "    \n",
    "    \n",
    "    sns.heatmap(cis_genes, cbar=False, cmap='inferno', square=True)\n",
    "    \n",
    "    h = max([len(u), len(d)])\n",
    "    h = int(h/10)\n",
    "    if h<1:\n",
    "        h=1\n",
    "    \n",
    "    if len(d)>0 and len(u)>0:\n",
    "        print(\"both up and down trans\")\n",
    "        fig, ax= plt.subplots(1,2, figsize=(20,h))\n",
    "\n",
    "        sns.heatmap(u, ax=ax[0], cbar=False, cmap='inferno')\n",
    "        sns.heatmap(d, ax=ax[1], cbar=False, cmap='inferno')\n",
    "    \n",
    "        ax[0].set_title(\"UP Trans genes\")\n",
    "        ax[1].set_title(\"DOWN Trans genes\")\n",
    "    \n",
    "    elif len(d)==0 and len(u)>0:\n",
    "  \n",
    "        fig, ax = plt.subplots(figsize=(10,h))\n",
    "        sns.heatmap(u, ax=ax, cbar=False, cmap='inferno')\n",
    "        ax.set_title(\"UP Trans genes\")\n",
    "    \n",
    "    elif len(d)>0 and len(u)==0:\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(10,h))\n",
    "        sns.heatmap(d, ax=ax, cbar=False, cmap='inferno')\n",
    "        ax.set_title(\"DOWN Trans genes\")\n",
    "                               \n",
    "    elif len(d)==0 and len(u)==0:\n",
    "        print(\"no significant trans-eQTL detected\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "6dd3cfd0-fb09-4c5f-88c0-fc0cd043e09d",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid = 'rs8073060'\n",
    "EA = 'A'\n",
    "OA = 'T'\n",
    "pos = 35548243\n",
    "chrom = 17\n",
    "\n",
    "x = Variant(rsid, chrom, pos, EA, OA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87f58cc4-84e6-4c2f-ae76-3e613d7cba88",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "987e538e-b971-4820-8ba8-d034b2972b04",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "f1b59285-bbb7-4a2b-aa1a-6a3863c0d50c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNPChr</th>\n",
       "      <th>SNPPos</th>\n",
       "      <th>AssessedAllele</th>\n",
       "      <th>OtherAllele</th>\n",
       "      <th>Zscore</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>GeneChr</th>\n",
       "      <th>GenePos</th>\n",
       "      <th>NrCohorts</th>\n",
       "      <th>NrSamples</th>\n",
       "      <th>FDR</th>\n",
       "      <th>BonferroniP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>49949</th>\n",
       "      <td>1.646800e-257</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>34.2785</td>\n",
       "      <td>ENSG00000105701</td>\n",
       "      <td>FKBP8</td>\n",
       "      <td>19</td>\n",
       "      <td>18648715</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.352000e-249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49950</th>\n",
       "      <td>4.050900e-242</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>33.2295</td>\n",
       "      <td>ENSG00000158856</td>\n",
       "      <td>EPB49</td>\n",
       "      <td>8</td>\n",
       "      <td>21923272</td>\n",
       "      <td>34</td>\n",
       "      <td>26110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.245500e-234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49951</th>\n",
       "      <td>1.181400e-200</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>30.2231</td>\n",
       "      <td>ENSG00000022840</td>\n",
       "      <td>RNF10</td>\n",
       "      <td>12</td>\n",
       "      <td>120993340</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.404700e-192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49952</th>\n",
       "      <td>1.290300e-149</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>26.0518</td>\n",
       "      <td>ENSG00000167671</td>\n",
       "      <td>UBXN6</td>\n",
       "      <td>19</td>\n",
       "      <td>4451407</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.626400e-141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49953</th>\n",
       "      <td>3.177200e-88</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>19.9124</td>\n",
       "      <td>ENSG00000167992</td>\n",
       "      <td>VWCE</td>\n",
       "      <td>11</td>\n",
       "      <td>61044329</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.467100e-80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50200</th>\n",
       "      <td>5.652200e-06</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4.5391</td>\n",
       "      <td>ENSG00000142784</td>\n",
       "      <td>WDTC1</td>\n",
       "      <td>1</td>\n",
       "      <td>27598058</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.030999</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50201</th>\n",
       "      <td>6.452300e-06</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4.5112</td>\n",
       "      <td>ENSG00000145423</td>\n",
       "      <td>SFRP2</td>\n",
       "      <td>4</td>\n",
       "      <td>154706008</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.034630</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50202</th>\n",
       "      <td>6.519500e-06</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-4.5087</td>\n",
       "      <td>ENSG00000198518</td>\n",
       "      <td>HIST1H4E</td>\n",
       "      <td>6</td>\n",
       "      <td>26205562</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.034925</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50203</th>\n",
       "      <td>7.734700e-06</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4.4724</td>\n",
       "      <td>ENSG00000105698</td>\n",
       "      <td>USF2</td>\n",
       "      <td>19</td>\n",
       "      <td>35765299</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.040104</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50204</th>\n",
       "      <td>8.249700e-06</td>\n",
       "      <td>rs8073060</td>\n",
       "      <td>17</td>\n",
       "      <td>35548243</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4.4587</td>\n",
       "      <td>ENSG00000196139</td>\n",
       "      <td>AKR1C3</td>\n",
       "      <td>10</td>\n",
       "      <td>5113713</td>\n",
       "      <td>35</td>\n",
       "      <td>31185</td>\n",
       "      <td>0.042174</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>256 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Pvalue        SNP  SNPChr    SNPPos AssessedAllele OtherAllele  \\\n",
       "49949  1.646800e-257  rs8073060      17  35548243              A           T   \n",
       "49950  4.050900e-242  rs8073060      17  35548243              A           T   \n",
       "49951  1.181400e-200  rs8073060      17  35548243              A           T   \n",
       "49952  1.290300e-149  rs8073060      17  35548243              A           T   \n",
       "49953   3.177200e-88  rs8073060      17  35548243              A           T   \n",
       "...              ...        ...     ...       ...            ...         ...   \n",
       "50200   5.652200e-06  rs8073060      17  35548243              A           T   \n",
       "50201   6.452300e-06  rs8073060      17  35548243              A           T   \n",
       "50202   6.519500e-06  rs8073060      17  35548243              A           T   \n",
       "50203   7.734700e-06  rs8073060      17  35548243              A           T   \n",
       "50204   8.249700e-06  rs8073060      17  35548243              A           T   \n",
       "\n",
       "        Zscore             Gene GeneSymbol GeneChr    GenePos  NrCohorts  \\\n",
       "49949  34.2785  ENSG00000105701      FKBP8      19   18648715         35   \n",
       "49950  33.2295  ENSG00000158856      EPB49       8   21923272         34   \n",
       "49951  30.2231  ENSG00000022840      RNF10      12  120993340         35   \n",
       "49952  26.0518  ENSG00000167671      UBXN6      19    4451407         35   \n",
       "49953  19.9124  ENSG00000167992       VWCE      11   61044329         35   \n",
       "...        ...              ...        ...     ...        ...        ...   \n",
       "50200   4.5391  ENSG00000142784      WDTC1       1   27598058         35   \n",
       "50201   4.5112  ENSG00000145423      SFRP2       4  154706008         35   \n",
       "50202  -4.5087  ENSG00000198518   HIST1H4E       6   26205562         35   \n",
       "50203   4.4724  ENSG00000105698       USF2      19   35765299         35   \n",
       "50204   4.4587  ENSG00000196139     AKR1C3      10    5113713         35   \n",
       "\n",
       "       NrSamples       FDR    BonferroniP  \n",
       "49949      31185  0.000000  3.352000e-249  \n",
       "49950      26110  0.000000  8.245500e-234  \n",
       "49951      31185  0.000000  2.404700e-192  \n",
       "49952      31185  0.000000  2.626400e-141  \n",
       "49953      31185  0.000000   6.467100e-80  \n",
       "...          ...       ...            ...  \n",
       "50200      31185  0.030999   1.000000e+00  \n",
       "50201      31185  0.034630   1.000000e+00  \n",
       "50202      31185  0.034925   1.000000e+00  \n",
       "50203      31185  0.040104   1.000000e+00  \n",
       "50204      31185  0.042174   1.000000e+00  \n",
       "\n",
       "[256 rows x 15 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans = pd.read_csv(get_paths(root=config.cbio_root)['eqtlgen_trans'], sep ='\\t')\n",
    "\n",
    "trans[trans.SNP==x.rsid]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb256657-28ea-402b-b385-939f3ba2647d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d04757ac-1ceb-4295-9752-4e10c09e78a8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47e2d756-95c2-497d-8b50-714404e7085f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31a96792-43c4-4740-ae59-6e3b82f37250",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57c62b84-58c9-4e1c-8112-a99d67a880e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf1fb4b-7c0c-43d9-88a0-99c014c67821",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d843738e-87fd-4f4e-b905-e55a60720a45",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbef52b8-2eaf-406f-a537-1be1b528fe8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb710053-382b-4591-ab8c-31db9b504c64",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2ad4dc5-cec8-4a8e-9cd7-e821b2aa9f3d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387a00a3-1612-4bed-a3da-22af37c5b96c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "768a9ec5-cbcf-4c6b-8bc3-b38011b6a35f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
