{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "21c9b44b-9032-488f-a68c-ef0af006f7fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import logging\n",
    "from ImmuneGWAS.helpers.getpaths import get_paths, get_sumstats_path\n",
    "from ImmuneGWAS.helpers import dbsnp, ldlink\n",
    "import ImmuneGWAS.resources.immune_GWAS as immune_GWAS\n",
    "from ImmuneGWAS import config\n",
    "from ImmuneGWAS.variant import Variant\n",
    "\n",
    "from ImmuneGWAS.resources.eqtlgen import eqtlgen_trans_LDblock_query \n",
    "from ImmuneGWAS.resources.eqtlgen import eqtlgen_cis_LDblock_query \n",
    "from ImmuneGWAS.resources.tokyo_eqtl import tokyo_eqtl_LDblock_query \n",
    "from ImmuneGWAS.resources.eqtl_cat import *\n",
    "from IPython.display import display_html \n",
    "from IPython.display import display,HTML\n",
    "\n",
    "from scipy.cluster import hierarchy\n",
    "\n",
    "from ImmuneGWAS.helpers import get_gene_symbol  # convert ensembl to gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e42f553-ec06-4d45-89ee-4132a647f95d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8f00a739-068a-4767-a88f-721b38c7cbea",
   "metadata": {},
   "source": [
    "Designing the lookup:\n",
    "\n",
    "1) Variant object calls trans_df lookup\n",
    "2) Harmonize according to EA/OA, raise exceptions for multi-allelic SNPs\n",
    "3) Split trans_df into up and down\n",
    "4) Return row-scaled heatmap df where cell ordering is hard-coded and rows are resorted for clustering\n",
    "5) Plotting function for heatmap df\n",
    "6) Get flow phenotypes\n",
    "7) Get LDTrait phenotypes\n",
    "8) Summarize phenotpyes into tables with betas (remove duplicate phenotypes etc)\n",
    "9) Get a single summary cis-eQTL table\n",
    "10) Return summary report - 2 heatmaps for up and down, cis-eQTLs, phenotypes summarized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d355c9bd-7599-46f1-9f16-dff7f9ab3a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid = \"rs1354034\"\n",
    "chrom = 3\n",
    "pos = 56815721\n",
    "EA = 'T'\n",
    "OA = 'C'\n",
    "\n",
    "x = Variant(rsid, chrom, pos, EA, OA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d1f08a26-1316-4e3f-b48b-95319325b017",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.get_gwas_phenotypes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "9cda26a3-fab4-42c9-91d1-84aacebf4c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def multi_column_df_display(list_dfs, cols=3):\n",
    "    html_table = \"<table style='width:100%; border:0px'>{content}</table>\"\n",
    "    html_row = \"<tr style='border:0px'>{content}</tr>\"\n",
    "    html_cell = \"<td style='width:{width}%;vertical-align:top;border:0px'>{{content}}</td>\"\n",
    "    html_cell = html_cell.format(width=100/cols)\n",
    "\n",
    "    cells = [ html_cell.format(content=df.to_html()) for df in list_dfs ]\n",
    "    cells += (cols - (len(list_dfs)%cols)) * [html_cell.format(content=\"\")] # pad\n",
    "    rows = [ html_row.format(content=\"\".join(cells[i:i+cols])) for i in range(0,len(cells),cols)]\n",
    "    display(HTML(html_table.format(content=\"\".join(rows))))\n",
    "\n",
    "def summary_trans(var: Variant):\n",
    "    \n",
    "    # retrieve trans-eQTLs\n",
    "    trans = eqtlgen_trans_LDblock_query(var)\n",
    "    trans.Zscore = trans.Zscore.astype(float)\n",
    "    \n",
    "    \n",
    "    # check EA for sign of beta\n",
    "    \n",
    "    if set(trans.AssessedAllele == x.EA) == {True}:\n",
    "        trans.Zscore = trans.Zscore * -1\n",
    "    \n",
    "    # split into up and down genes\n",
    "    up = trans[trans.Zscore>0]\n",
    "    down = trans[trans.Zscore<0]\n",
    "    up = up.Gene\n",
    "    down = down.Gene\n",
    "    \n",
    "    # dictionary to map ensembl to gene name\n",
    "    mapper = dict(zip(trans.Gene, trans.GeneSymbol))\n",
    "    \n",
    "    #get ge matrix from ImmunexUT\n",
    "    paths = get_paths(root='/media/')\n",
    "    ge = pd.read_csv(paths['ge_tokyo'], sep='\\t')\n",
    "    ge = ge.set_index(\"Gene_id\")\n",
    "    ge = ge.drop(\"Gene_name\", axis=1)\n",
    "    ge = ge.drop_duplicates()\n",
    "    \n",
    "    # get indices for trans-eQTL matrix\n",
    "    # note any missing genes will be dropped\n",
    "    \n",
    "    u = [i for i in up if i in ge.index]\n",
    "    u = set(u)\n",
    "    d = [i for i in down if i in ge.index]\n",
    "    d = set(d)\n",
    "    \n",
    "    # filter gene expression matrix into up and down genes\n",
    "    u = ge.loc[u]\n",
    "    d = ge.loc[d]\n",
    "    \n",
    "    # rename ensembl IDs to gene names\n",
    "    u.index = u.index.map(mapper)\n",
    "    d.index = d.index.map(mapper)\n",
    "\n",
    "    # hard-code order of cell types according to gene expression similarity on whole GE matrix\n",
    "    \n",
    "    ordered_cols = ['Plasmablast', 'LDG', 'Neu', 'pDC', 'mDC', 'CL_Mono', 'Int_Mono',\n",
    "           'CD16p_Mono', 'NC_Mono', 'Naive_B', 'USM_B', 'DN_B', 'SM_B', 'NK',\n",
    "           'Mem_CD8', 'EM_CD8', 'TEMRA_CD8', 'Naive_CD8', 'Fr_I_nTreg',\n",
    "           'Naive_CD4', 'Fr_III_T', 'Fr_II_eTreg', 'CM_CD8', 'Th2', 'Tfh', 'Th17',\n",
    "           'Mem_CD4', 'Th1']\n",
    "    \n",
    "    # normalize up and down genes\n",
    "\n",
    "    def standard_scale(x):\n",
    "        return (x-x.min())/x.max()\n",
    "\n",
    "    u = u.apply(standard_scale, axis=1)\n",
    "    d = d.apply(standard_scale, axis=1)\n",
    "\n",
    "    # reorder columns\n",
    "    \n",
    "    u = u.loc[:, ordered_cols]\n",
    "    d = d.loc[:, ordered_cols]\n",
    "    \n",
    "    # reoder rows by dendrogram\n",
    "    if len(u)>1:\n",
    "        Z = hierarchy.linkage(u, optimal_ordering=True, method='ward')\n",
    "        dendro = hierarchy.dendrogram(Z, no_plot=True)\n",
    "    \n",
    "        u = u.iloc[dendro['leaves'],:]\n",
    "    \n",
    "    if len(d)>1:\n",
    "        Z = hierarchy.linkage(d, optimal_ordering=True, method='ward')\n",
    "        dendro = hierarchy.dendrogram(Z, no_plot=True)\n",
    "        d = d.iloc[dendro['leaves'],:]\n",
    "    # collect all my cis-eqtls\n",
    "\n",
    "    lookup_cis(x)\n",
    "    lookup_tokyo(x)\n",
    "    eqtl_catalogue_LDblock_query_type_restricted_multitype(x)\n",
    "\n",
    "    # now all of these are in the Results object\n",
    "\n",
    "    gwas = ldlink.ldtrait(x)\n",
    "\n",
    "    tokyo = x.results.tokyo_eqtl()\n",
    "    tokyo = tokyo[['Gene_name','Variant_ID', 'Backward_slope', 'Backward_P', 'cell_type']]\n",
    "    tokyo = tokyo.drop_duplicates()\n",
    "\n",
    "    cat = x.results.eqtl_cat()\n",
    "    cat = cat[['gene_symbol', 'variant', 'z', 'study', 'cell_type']]\n",
    "    \n",
    "    \n",
    "    gwas_traits = set(x.results.ldtrait()['GWAS Trait'].tolist())\n",
    "    \n",
    "    \n",
    "    cisgen = x.results.eqtlgen_cis_df\n",
    "    cisgen = cisgen[['SNP', 'GeneSymbol', 'Zscore']]\n",
    "    \n",
    "    \n",
    "    # get the cis-eqtl genes to plot a heatmap of the cis-eQTL genes\n",
    "    \n",
    "    a = x.results.eqtlgen_cis_df['Gene']\n",
    "\n",
    "    b = x.results.tokyo_eqtl_df['Gene_id'].apply(lambda x:x.split(\".\")[0])\n",
    "\n",
    "    c = x.results.eqtl_cat_df['molecular_trait_id']\n",
    "    \n",
    "    cis_genes = list(set(pd.concat([a,b,c]).tolist()))\n",
    "    cis_genes = [i for i in cis_genes if i in ge.index]\n",
    "    cis_genes = ge.loc[cis_genes]\n",
    "    cis_genes = cis_genes.apply(standard_scale, axis=1)\n",
    "    cis_genes = cis_genes.loc[:, ordered_cols]\n",
    "    cis_genes.index = cis_genes.index.map(get_gene_symbol)\n",
    "    \n",
    "    list_dfs = [tokyo, cat, cisgen]\n",
    "\n",
    "    display(tokyo)\n",
    "    display(cat)\n",
    "    display(cisgen)\n",
    "    print(\"GWAS Catalog\\n\")\n",
    "    display(gwas_traits)\n",
    "    print(\"Flow phenos\\n\")\n",
    "    display(x.get_gwas_phenotypes())\n",
    "    \n",
    "    sns.heatmap(cis_genes, cbar=False, cmap='inferno', square=True)\n",
    "    \n",
    "    h = max([len(u), len(d)])\n",
    "    h = int(h/10)\n",
    "    fig, ax= plt.subplots(1,2, figsize=(20,h))\n",
    "\n",
    "    sns.heatmap(u, ax=ax[0], cbar=False, cmap='inferno')\n",
    "    sns.heatmap(d, ax=ax[1], cbar=False, cmap='inferno')\n",
    "    \n",
    "    ax[0].set_title(\"UP Trans genes\")\n",
    "    ax[1].set_title(\"DOWN Trans genes\")\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "6dd3cfd0-fb09-4c5f-88c0-fc0cd043e09d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>CHR</th>\n",
       "      <th>TSS_position</th>\n",
       "      <th>Number_of_variants_cis</th>\n",
       "      <th>Variant_ID</th>\n",
       "      <th>OA</th>\n",
       "      <th>EA</th>\n",
       "      <th>Variant_CHR</th>\n",
       "      <th>Variant_position_start</th>\n",
       "      <th>Variant_position_end</th>\n",
       "      <th>Rank_of_association</th>\n",
       "      <th>Forward_nominal_P</th>\n",
       "      <th>Forward_slope</th>\n",
       "      <th>Backward_P</th>\n",
       "      <th>Backward_slope</th>\n",
       "      <th>int_chrom</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000163947.11</td>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>57079329</td>\n",
       "      <td>5997</td>\n",
       "      <td>rs1354034</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr3</td>\n",
       "      <td>56815721</td>\n",
       "      <td>56815721</td>\n",
       "      <td>0</td>\n",
       "      <td>1.75008e-05</td>\n",
       "      <td></td>\n",
       "      <td>1.75008e-05</td>\n",
       "      <td>0.165671</td>\n",
       "      <td>3</td>\n",
       "      <td>Plasmablast</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene_id Gene_name   CHR TSS_position Number_of_variants_cis  \\\n",
       "0  ENSG00000163947.11   ARHGEF3  chr3     57079329                   5997   \n",
       "\n",
       "  Variant_ID OA EA Variant_CHR Variant_position_start Variant_position_end  \\\n",
       "0  rs1354034  T  C        chr3               56815721             56815721   \n",
       "\n",
       "  Rank_of_association Forward_nominal_P Forward_slope   Backward_P  \\\n",
       "0                   0       1.75008e-05                1.75008e-05   \n",
       "\n",
       "  Backward_slope int_chrom    cell_type  \n",
       "0       0.165671         3  Plasmablast  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.results.tokyo_eqtl_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "d04757ac-1ceb-4295-9752-4e10c09e78a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNPChr</th>\n",
       "      <th>SNPPos</th>\n",
       "      <th>AssessedAllele</th>\n",
       "      <th>OtherAllele</th>\n",
       "      <th>Zscore</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>GeneChr</th>\n",
       "      <th>GenePos</th>\n",
       "      <th>NrCohorts</th>\n",
       "      <th>NrSamples</th>\n",
       "      <th>FDR</th>\n",
       "      <th>BonferroniP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7.2033e-07</td>\n",
       "      <td>rs10004195</td>\n",
       "      <td>4</td>\n",
       "      <td>38783103</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4.956</td>\n",
       "      <td>ENSG00000132185</td>\n",
       "      <td>FCRLA</td>\n",
       "      <td>1</td>\n",
       "      <td>161680452</td>\n",
       "      <td>35</td>\n",
       "      <td>31484</td>\n",
       "      <td>0.0052005816439996</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.7064e-06</td>\n",
       "      <td>rs10004195</td>\n",
       "      <td>4</td>\n",
       "      <td>38783103</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4.6273</td>\n",
       "      <td>ENSG00000144645</td>\n",
       "      <td>OSBPL10</td>\n",
       "      <td>3</td>\n",
       "      <td>31909227</td>\n",
       "      <td>35</td>\n",
       "      <td>31484</td>\n",
       "      <td>0.0219289541575709</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7.6411e-06</td>\n",
       "      <td>rs10004195</td>\n",
       "      <td>4</td>\n",
       "      <td>38783103</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>4.4752</td>\n",
       "      <td>ENSG00000169429</td>\n",
       "      <td>IL8</td>\n",
       "      <td>4</td>\n",
       "      <td>74607828</td>\n",
       "      <td>32</td>\n",
       "      <td>25446</td>\n",
       "      <td>0.0396986665127287</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.2593e-07</td>\n",
       "      <td>rs4833095</td>\n",
       "      <td>4</td>\n",
       "      <td>38798089</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5.1078</td>\n",
       "      <td>ENSG00000132185</td>\n",
       "      <td>FCRLA</td>\n",
       "      <td>1</td>\n",
       "      <td>161680452</td>\n",
       "      <td>35</td>\n",
       "      <td>30962</td>\n",
       "      <td>0.0024900587010036</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0063e-06</td>\n",
       "      <td>rs17616434</td>\n",
       "      <td>4</td>\n",
       "      <td>38811255</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4.8905</td>\n",
       "      <td>ENSG00000132185</td>\n",
       "      <td>FCRLA</td>\n",
       "      <td>1</td>\n",
       "      <td>161680452</td>\n",
       "      <td>35</td>\n",
       "      <td>30962</td>\n",
       "      <td>0.0069285947179654</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Pvalue         SNP SNPChr    SNPPos AssessedAllele OtherAllele  Zscore  \\\n",
       "0  7.2033e-07  rs10004195      4  38783103              A           T   4.956   \n",
       "1  3.7064e-06  rs10004195      4  38783103              A           T  4.6273   \n",
       "2  7.6411e-06  rs10004195      4  38783103              A           T  4.4752   \n",
       "0  3.2593e-07   rs4833095      4  38798089              C           T  5.1078   \n",
       "0  1.0063e-06  rs17616434      4  38811255              C           T  4.8905   \n",
       "\n",
       "              Gene GeneSymbol GeneChr    GenePos NrCohorts NrSamples  \\\n",
       "0  ENSG00000132185      FCRLA       1  161680452        35     31484   \n",
       "1  ENSG00000144645    OSBPL10       3   31909227        35     31484   \n",
       "2  ENSG00000169429        IL8       4   74607828        32     25446   \n",
       "0  ENSG00000132185      FCRLA       1  161680452        35     30962   \n",
       "0  ENSG00000132185      FCRLA       1  161680452        35     30962   \n",
       "\n",
       "                  FDR BonferroniP  \n",
       "0  0.0052005816439996         1.0  \n",
       "1  0.0219289541575709         1.0  \n",
       "2  0.0396986665127287         1.0  \n",
       "0  0.0024900587010036         1.0  \n",
       "0  0.0069285947179654         1.0  "
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans = eqtlgen_trans_LDblock_query(x)\n",
    "\n",
    "trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "57c62b84-58c9-4e1c-8112-a99d67a880e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rs10004195'"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.rsid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "ebf1fb4b-7c0c-43d9-88a0-99c014c67821",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A'"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.EA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "d843738e-87fd-4f4e-b905-e55a60720a45",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'T'"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.OA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "fbef52b8-2eaf-406f-a537-1be1b528fe8f",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "ERROR: LDblock EA C does not correspond to either Tokyo EAGC or Tokyo OA G",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-170-365192240ba7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msummary_trans\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-169-7a6a5619e4df>\u001b[0m in \u001b[0;36msummary_trans\u001b[0;34m(var)\u001b[0m\n\u001b[1;32m     89\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m     \u001b[0mlookup_cis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 91\u001b[0;31m     \u001b[0mlookup_tokyo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     92\u001b[0m     \u001b[0meqtl_catalogue_LDblock_query_type_restricted_multitype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     93\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Projects/ImmuneGWAS/ImmuneGWAS/resources/tokyo_eqtl.py\u001b[0m in \u001b[0;36mtokyo_eqtl_LDblock_query\u001b[0;34m(variant_object)\u001b[0m\n\u001b[1;32m    113\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mvariant_pos_list\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvariant_positions_list_of_lists\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Iterate over the LDblock, make a query for each SNP\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    114\u001b[0m             \u001b[0;31m# Create a DataFrame with all the matches for the variant at the position (list[0], list[1])\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 115\u001b[0;31m             \u001b[0mvariant_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msingle_tokyo_eqtl_query\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvariant_pos_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariant_pos_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mEA\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvariant_pos_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    116\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mvariant_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    117\u001b[0m                 \u001b[0;31m# Companion message for message coming from single_tokyo_eqtl_query.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Projects/ImmuneGWAS/ImmuneGWAS/resources/tokyo_eqtl.py\u001b[0m in \u001b[0;36msingle_tokyo_eqtl_query\u001b[0;34m(chromosome, position, EA)\u001b[0m\n\u001b[1;32m     77\u001b[0m                 \u001b[0mconcatenated_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mForward_slope\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconcatenated_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mForward_slope\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     78\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 79\u001b[0;31m                 raise ValueError(f\"ERROR: LDblock EA {EA} does not correspond to either Tokyo EA\"\n\u001b[0m\u001b[1;32m     80\u001b[0m                                  f\"{concatenated_df['EA'].iloc[0]} or Tokyo OA {concatenated_df['OA'].iloc[0]}\")\n\u001b[1;32m     81\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mconcatenated_df\u001b[0m  \u001b[0;31m# Results from every cell type for a single SNP\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: ERROR: LDblock EA C does not correspond to either Tokyo EAGC or Tokyo OA G"
     ]
    }
   ],
   "source": [
    "summary_trans(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "eb710053-382b-4591-ab8c-31db9b504c64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'eqtl_cat': '/media/cbio3/data/eQTL_DB/',\n",
       " 'dbsnp': '/media/cbio3/data/dbSNP/GCF_000001405.39.gz',\n",
       " 'ensembl': '/media/cbio3/data/ensembl_biomart/gene_aliases.json',\n",
       " 'eqtl_tokyo': '/media/cbio3/projects/Zain_2021/ImmuNEXT_Japan_Cell2021/eQTL_summarystats_with_alleles/',\n",
       " 'ge_tokyo': '/media/cbio3/projects/Zain_2021/ImmunexUT_GE/E-GEAD-397.processed/antton_reprocessing/mean_table_TPM.txt',\n",
       " 'eqtlgen_cis': '/media/cbio3/projects/Zain_2021/eQTLgen/data/2019-12-11-cis-eQTLsFDR0.05-ProbeLevel-CohortInfoRemoved-BonferroniAdded_REsorted_hg38.txt.gz',\n",
       " 'eqtlgen_trans': '/media/cbio3/projects/Zain_2021/eQTLgen/data/2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded_sorted_hg38.txt.gz',\n",
       " 'tokyo_alleles': '/media/cbio3/projects/Zain_2021/ImmuNEXT_Japan_Cell2021/metadata/full_allele_defs.txt',\n",
       " 'immuneGWAS': '/media/cbio3/projects/antton/Immune_cell_GWAS/data/hits_only_table_hg38.txt.gz'}"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2ad4dc5-cec8-4a8e-9cd7-e821b2aa9f3d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387a00a3-1612-4bed-a3da-22af37c5b96c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "768a9ec5-cbcf-4c6b-8bc3-b38011b6a35f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
