{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "21c9b44b-9032-488f-a68c-ef0af006f7fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import logging\n",
    "from ImmuneGWAS.helpers.getpaths import get_paths, get_sumstats_path\n",
    "from ImmuneGWAS.helpers import dbsnp, ldlink\n",
    "import ImmuneGWAS.resources.immune_GWAS as immune_GWAS\n",
    "from ImmuneGWAS import config\n",
    "from ImmuneGWAS.variant import Variant\n",
    "\n",
    "from ImmuneGWAS.resources.eqtlgen import eqtlgen_trans_LDblock_query \n",
    "from ImmuneGWAS.resources.eqtlgen import eqtlgen_cis_LDblock_query \n",
    "from ImmuneGWAS.resources.tokyo_eqtl import tokyo_eqtl_LDblock_query\n",
    "from ImmuneGWAS.resources.eqtl_cat import *\n",
    "from IPython.display import display_html \n",
    "from IPython.display import display,HTML\n",
    "\n",
    "from scipy.cluster import hierarchy\n",
    "\n",
    "from ImmuneGWAS.helpers import get_gene_symbol  # convert ensembl to gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7e42f553-ec06-4d45-89ee-4132a647f95d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ldpruned_hits = \"/media/cbio3/projects/antton/Immune_cell_GWAS/data/hits_only_table_hg38_LDprunned.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f00a739-068a-4767-a88f-721b38c7cbea",
   "metadata": {},
   "source": [
    "Designing the lookup:\n",
    "\n",
    "1) Variant object calls trans_df lookup\n",
    "2) Harmonize according to EA/OA, raise exceptions for multi-allelic SNPs\n",
    "3) Split trans_df into up and down\n",
    "4) Return row-scaled heatmap df where cell ordering is hard-coded and rows are resorted for clustering\n",
    "5) Plotting function for heatmap df\n",
    "6) Get flow phenotypes\n",
    "7) Get LDTrait phenotypes\n",
    "8) Summarize phenotpyes into tables with betas (remove duplicate phenotypes etc)\n",
    "9) Get a single summary cis-eQTL table\n",
    "10) Return summary report - 2 heatmaps for up and down, cis-eQTLs, phenotypes summarized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d355c9bd-7599-46f1-9f16-dff7f9ab3a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid = \"rs1354034\"\n",
    "chrom = 3\n",
    "pos = 56815721\n",
    "EA = 'T'\n",
    "OA = 'C'\n",
    "\n",
    "x = Variant(rsid, chrom, pos, EA, OA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f08a26-1316-4e3f-b48b-95319325b017",
   "metadata": {},
   "outputs": [],
   "source": [
    "x.results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "9cda26a3-fab4-42c9-91d1-84aacebf4c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "def summary_trans(x: Variant):\n",
    "    \n",
    "    # retrieve trans-eQTLs\n",
    "    eqtlgen_trans_LDblock_query(x)\n",
    "    trans = x.results.eqtlgen_trans()\n",
    "    trans.Zscore = trans.Zscore.astype(float)\n",
    "    \n",
    "    \n",
    "    # check EA for sign of beta\n",
    "    \n",
    "    if set(trans.AssessedAllele == x.EA) == {True}:\n",
    "        trans.Zscore = trans.Zscore.astype(float) * -1\n",
    "    \n",
    "    # split into up and down genes\n",
    "    up = trans[trans.Zscore>0]\n",
    "    down = trans[trans.Zscore<0]\n",
    "    up = up.Gene\n",
    "    down = down.Gene\n",
    "    \n",
    "    \n",
    "    #get ge matrix from ImmunexUT\n",
    "    # Get ge matrix from ImmunexUT\n",
    "    paths = get_paths(root=config.cbio_root)\n",
    "    ge = pd.read_csv(paths['ge_tokyo'], sep='\\t')\n",
    "    ge = ge.set_index(\"Gene_id\")\n",
    "    ge = ge.drop(\"Gene_name\", axis=1)\n",
    "    ge = ge.drop_duplicates()\n",
    "    \n",
    "    # get indices for trans-eQTL matrix\n",
    "    # note any missing genes will be dropped\n",
    "    \n",
    "    u = [i for i in up if i in ge.index]\n",
    "    u = set(u)\n",
    "    d = [i for i in down if i in ge.index]\n",
    "    d = set(d)\n",
    "    \n",
    "    # filter gene expression matrix into up and down genes\n",
    "    u = ge.loc[u]\n",
    "    d = ge.loc[d]\n",
    "    \n",
    "    # rename ensembl IDs to gene names\n",
    "    u.index = u.index.map(mapper)\n",
    "    d.index = d.index.map(mapper)\n",
    "\n",
    "    # hard-code order of cell types according to gene expression similarity on whole GE matrix\n",
    "    \n",
    "    ordered_cols = ['Plasmablast', 'LDG', 'Neu', 'pDC', 'mDC', 'CL_Mono', 'Int_Mono',\n",
    "           'CD16p_Mono', 'NC_Mono', 'Naive_B', 'USM_B', 'DN_B', 'SM_B', 'NK',\n",
    "           'Mem_CD8', 'EM_CD8', 'TEMRA_CD8', 'Naive_CD8', 'Fr_I_nTreg',\n",
    "           'Naive_CD4', 'Fr_III_T', 'Fr_II_eTreg', 'CM_CD8', 'Th2', 'Tfh', 'Th17',\n",
    "           'Mem_CD4', 'Th1']\n",
    "    \n",
    "    # normalize up and down genes\n",
    "\n",
    "    def standard_scale(x):\n",
    "        return (x-x.min())/x.max()\n",
    "\n",
    "    u = u.apply(standard_scale, axis=1)\n",
    "    d = d.apply(standard_scale, axis=1)\n",
    "\n",
    "    # reorder columns\n",
    "    \n",
    "    u = u.loc[:, ordered_cols]\n",
    "    d = d.loc[:, ordered_cols]\n",
    "    \n",
    "    # reoder rows by dendrogram\n",
    "    if len(u)>1:\n",
    "        Z = hierarchy.linkage(u, optimal_ordering=True, method='ward')\n",
    "        dendro = hierarchy.dendrogram(Z, no_plot=True)\n",
    "    \n",
    "        u = u.iloc[dendro['leaves'],:]\n",
    "    \n",
    "    if len(d)>1:\n",
    "        Z = hierarchy.linkage(d, optimal_ordering=True, method='ward')\n",
    "        dendro = hierarchy.dendrogram(Z, no_plot=True)\n",
    "        d = d.iloc[dendro['leaves'],:]\n",
    "    \n",
    "    # collect all my cis-eqtls\n",
    "\n",
    "    eqtlgen_cis_LDblock_query(x)\n",
    "    tokyo_eqtl_LDblock_query(x)\n",
    "    eqtl_catalogue_LDblock_query_type_restricted_multitype(x)\n",
    "\n",
    "    # now all of these are in the Results object\n",
    "\n",
    "    gwas = ldlink.ldtrait(x)\n",
    "\n",
    "    # first try to filter by the same SNP as in Variant object, otherwise filter down to one rsid\n",
    "    \n",
    "    tokyo = x.results.tokyo_eqtl()[x.results.tokyo_eqtl()['Variant_ID']==x.rsid]\n",
    "    \n",
    "    if len(tokyo)>0:\n",
    "        tokyo = tokyo[['Gene_name','Variant_ID', 'Backward_slope', 'Backward_P', 'cell_type']]\n",
    "        tokyo = tokyo.sort_values(by='Gene_name')\n",
    "    \n",
    "    elif len(tokyo)==0:\n",
    "        tokyo = x.results.tokyo_eqtl()\n",
    "        rsid = tokyo['Variant ID'][0]\n",
    "        \n",
    "\n",
    "    \n",
    "    cat = x.results.eqtl_cat()\n",
    "    if len(cat)>0:\n",
    "        cat = cat[cat.position==str(x.pos)]\n",
    "        cat = cat[['gene_symbol', 'z', 'study', 'cell_type']]\n",
    "    \n",
    "    \n",
    "    gwas_traits = set(x.results.ldtrait()['GWAS Trait'].tolist())\n",
    "    \n",
    "    \n",
    "    \n",
    "    cisgen = x.results.eqtlgen_cis()\n",
    "    cisgen = cisgen[cisgen.SNP==x.rsid]\n",
    "    if len(cisgen)>0:\n",
    "        cisgen = cisgen[['SNP', 'Zscore', 'Pvalue', 'GeneSymbol']].sort_values(by='GeneSymbol')\n",
    "    elif len(csigen)==0:\n",
    "        cisgen = x.results.eqtlgen_cis()\n",
    "        rsid = cisgen.SNP[0]\n",
    "        cisgen = cisgen[cisgen.SNP==rsid]\n",
    "        cisgen = cisgen[['SNP', 'Zscore', 'Pvalue', 'GeneSymbol']].sort_values(by='GeneSymbol')\n",
    "        \n",
    "    # get the cis-eqtl genes to plot a heatmap of the cis-eQTL genes\n",
    "    \n",
    "    if len(x.results.eqtlgen_cis())>0:\n",
    "        a = x.results.eqtlgen_cis_df['Gene'].tolist()\n",
    "    else:\n",
    "        a = []\n",
    "\n",
    "    if len(x.results.tokyo_eqtl())>0:\n",
    "        b = x.results.tokyo_eqtl_df['Gene_id'].apply(lambda x:x.split(\".\")[0]).tolist()\n",
    "    else:\n",
    "        b = []\n",
    "    \n",
    "    if len(x.results.eqtl_cat())>0:\n",
    "        c = x.results.eqtl_cat_df['molecular_trait_id'].tolist()\n",
    "    else:\n",
    "        c = []\n",
    "    \n",
    "    cis_genes = a + b + c\n",
    "    cis_genes = [i for i in cis_genes if i in ge.index]\n",
    "    cis_genes = list(set(cis_genes))\n",
    "    cis_genes = ge.loc[cis_genes]\n",
    "    cis_genes = cis_genes.apply(standard_scale, axis=1)\n",
    "    cis_genes = cis_genes.loc[:, ordered_cols]\n",
    "    cis_genes.index = cis_genes.index.map(get_gene_symbol)\n",
    "    \n",
    "\n",
    "    display(tokyo)\n",
    "    display(cat)\n",
    "    display(cisgen)\n",
    "    print(\"GWAS Catalog\\n\")\n",
    "    display(gwas_traits)\n",
    "    print(\"Flow phenos\\n\")\n",
    "    display(x.get_gwas_phenotypes())\n",
    "    \n",
    "    \n",
    "    sns.heatmap(cis_genes, cbar=False, cmap='inferno', square=True)\n",
    "    \n",
    "    h = max([len(u), len(d)])\n",
    "    h = int(h/10)\n",
    "    if h<1:\n",
    "        h=1\n",
    "    \n",
    "    if len(d)>0 and len(u)>0:\n",
    "        print(\"both up and down trans\")\n",
    "        fig, ax= plt.subplots(1,2, figsize=(20,h))\n",
    "\n",
    "        sns.heatmap(u, ax=ax[0], cbar=False, cmap='inferno')\n",
    "        sns.heatmap(d, ax=ax[1], cbar=False, cmap='inferno')\n",
    "    \n",
    "        ax[0].set_title(\"UP Trans genes\")\n",
    "        ax[1].set_title(\"DOWN Trans genes\")\n",
    "    \n",
    "    elif len(d)==0 and len(u)>0:\n",
    "  \n",
    "        fig, ax = plt.subplots(figsize=(10,h))\n",
    "        sns.heatmap(u, ax=ax, cbar=False, cmap='inferno')\n",
    "        ax.set_title(\"UP Trans genes\")\n",
    "    \n",
    "    elif len(d)>0 and len(u)==0:\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(10,h))\n",
    "        sns.heatmap(d, ax=ax, cbar=False, cmap='inferno')\n",
    "        ax.set_title(\"DOWN Trans genes\")\n",
    "                               \n",
    "    elif len(d)==0 and len(u)==0:\n",
    "        print(\"no significant trans-eQTL detected\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6dd3cfd0-fb09-4c5f-88c0-fc0cd043e09d",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid = 'rs8073060'\n",
    "EA = 'A'\n",
    "OA = 'T'\n",
    "pos = 35548243\n",
    "chrom = 17\n",
    "\n",
    "x = Variant(rsid, chrom, pos, EA, OA)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "87f58cc4-84e6-4c2f-ae76-3e613d7cba88",
   "metadata": {},
   "outputs": [],
   "source": [
    "paths = get_paths(root=config.cbio_root)\n",
    "ge = pd.read_csv(paths['ge_tokyo'], sep='\\t')\n",
    "ge = pd.read_csv(paths['ge_tokyo'], sep='\\t')\n",
    "ge = ge.set_index(\"Gene_id\")\n",
    "ge = ge.drop(\"Gene_name\", axis=1)\n",
    "ge = ge.drop_duplicates()\n",
    "\n",
    "a = x.results.eqtlgen_cis_df['Gene'].tolist()\n",
    "b = x.results.tokyo_eqtl_df['Gene_id'].apply(lambda x:x.split(\".\")[0]).tolist()\n",
    "c = []\n",
    "\n",
    "cis_genes = a + b + c\n",
    "cis_genes = [i for i in cis_genes if i in ge.index]\n",
    "cis_genes = list(set(cis_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "987e538e-b971-4820-8ba8-d034b2972b04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSG00000172716',\n",
       " 'ENSG00000255987',\n",
       " 'ENSG00000154760',\n",
       " 'ENSG00000236320',\n",
       " 'ENSG00000205045',\n",
       " 'ENSG00000005156',\n",
       " 'ENSG00000267369',\n",
       " 'ENSG00000108733',\n",
       " 'ENSG00000267074',\n",
       " 'ENSG00000267364',\n",
       " 'ENSG00000270806']"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cis_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "f1b59285-bbb7-4a2b-aa1a-6a3863c0d50c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000172716\n",
      "SLFN11\n",
      "ENSG00000255987\n",
      "TOMM20P2\n",
      "ENSG00000154760\n",
      "SLFN13\n",
      "ENSG00000236320\n",
      "SLFN14\n",
      "ENSG00000205045\n",
      "SLFN12L\n",
      "ENSG00000005156\n",
      "LIG3\n",
      "ENSG00000267369\n",
      "TAF5LP1\n",
      "ENSG00000108733\n",
      "PEX12\n",
      "ENSG00000267074\n",
      "ENSG00000267074 is missing!\n",
      "ENSG00000267364\n",
      "ENSG00000267364 is missing!\n",
      "ENSG00000270806\n",
      "C17orf50\n"
     ]
    }
   ],
   "source": [
    "for i in cis_genes:\n",
    "    print(i)\n",
    "    try:\n",
    "        print(get_gene_symbol(i))\n",
    "    except:\n",
    "        print(i + \" is missing!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb256657-28ea-402b-b385-939f3ba2647d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d04757ac-1ceb-4295-9752-4e10c09e78a8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e2d756-95c2-497d-8b50-714404e7085f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31a96792-43c4-4740-ae59-6e3b82f37250",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57c62b84-58c9-4e1c-8112-a99d67a880e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf1fb4b-7c0c-43d9-88a0-99c014c67821",
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_catalogue_LDblock_query_type_restricted_multitype(x)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d843738e-87fd-4f4e-b905-e55a60720a45",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbef52b8-2eaf-406f-a537-1be1b528fe8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb710053-382b-4591-ab8c-31db9b504c64",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2ad4dc5-cec8-4a8e-9cd7-e821b2aa9f3d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387a00a3-1612-4bed-a3da-22af37c5b96c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "768a9ec5-cbcf-4c6b-8bc3-b38011b6a35f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}