{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "21c9b44b-9032-488f-a68c-ef0af006f7fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import subprocess\n",
    "import logging\n",
    "from ImmuneGWAS.helpers.getpaths import get_paths, get_sumstats_path\n",
    "from ImmuneGWAS.helpers import dbsnp, ldlink\n",
    "import ImmuneGWAS.resources.immune_GWAS as immune_GWAS\n",
    "from ImmuneGWAS import config\n",
    "from ImmuneGWAS.variant import Variant\n",
    "\n",
    "from ImmuneGWAS.resources.eqtlgen import eqtlgen_trans_LDblock_query as lookup_trans\n",
    "from ImmuneGWAS.resources.eqtlgen import eqtlgen_cis_LDblock_query as lookup_cis\n",
    "from ImmuneGWAS.resources.tokyo_eqtl import tokyo_eqtl_LDblock_query as lookup_tokyo\n",
    "from ImmuneGWAS.resources.eqtl_cat import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e42f553-ec06-4d45-89ee-4132a647f95d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8f00a739-068a-4767-a88f-721b38c7cbea",
   "metadata": {},
   "source": [
    "Designing the lookup:\n",
    "\n",
    "1) Variant object calls trans_df lookup\n",
    "2) Harmonize according to EA/OA, raise exceptions for multi-allelic SNPs\n",
    "3) Split trans_df into up and down\n",
    "4) Return row-scaled heatmap df where cell ordering is hard-coded and rows are resorted for clustering\n",
    "5) Plotting function for heatmap df\n",
    "6) Get flow phenotypes\n",
    "7) Get LDTrait phenotypes\n",
    "8) Summarize phenotpyes into tables with betas (remove duplicate phenotypes etc)\n",
    "9) Get a single summary cis-eQTL table\n",
    "10) Return summary report - 2 heatmaps for up and down, cis-eQTLs, phenotypes summarized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d355c9bd-7599-46f1-9f16-dff7f9ab3a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid = \"rs1354034\"\n",
    "chrom = 3\n",
    "pos = 56815721\n",
    "EA = 'T'\n",
    "OA = 'C'\n",
    "\n",
    "x = Variant(rsid, chrom, pos, EA, OA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cda26a3-fab4-42c9-91d1-84aacebf4c90",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "05b3efdd-1c8d-4f18-963e-e6609663f043",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get trans DF\n",
    "\n",
    "trans = lookup_trans(x)\n",
    "trans.Zscore = trans.Zscore.astype(float)\n",
    "    \n",
    "\n",
    "# check EA for sign of beta\n",
    "if set(trans.AssessedAllele == x.EA) == {True}:\n",
    "    trans.Zscore = trans.Zscore * -1\n",
    "\n",
    "up = trans[trans.Zscore>0]\n",
    "down = trans[trans.Zscore<0]\n",
    "\n",
    "up = up.Gene\n",
    "down = down.Gene\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# fig, ax = plt.subplots()\n",
    "\n",
    "mapper = dict(zip(trans.Gene, trans.GeneSymbol))\n",
    "\n",
    "paths = get_paths(root='/media/')\n",
    "\n",
    "ge = pd.read_csv(paths['ge_tokyo'], sep='\\t')\n",
    "\n",
    "ge = ge.set_index(\"Gene_id\")\n",
    "\n",
    "ge = ge.drop(\"Gene_name\", axis=1)\n",
    "\n",
    "\n",
    "ge = ge.drop_duplicates()\n",
    "\n",
    "u = [i for i in up if i in ge.index]\n",
    "u = set(u)\n",
    "d = [i for i in down if i in ge.index]\n",
    "d = set(d)\n",
    "u = ge.loc[u]\n",
    "d = ge.loc[d]\n",
    "\n",
    "u.index = u.index.map(mapper)\n",
    "d.index = d.index.map(mapper)\n",
    "\n",
    "ordered_cols = ['Plasmablast', 'LDG', 'Neu', 'pDC', 'mDC', 'CL_Mono', 'Int_Mono',\n",
    "       'CD16p_Mono', 'NC_Mono', 'Naive_B', 'USM_B', 'DN_B', 'SM_B', 'NK',\n",
    "       'Mem_CD8', 'EM_CD8', 'TEMRA_CD8', 'Naive_CD8', 'Fr_I_nTreg',\n",
    "       'Naive_CD4', 'Fr_III_T', 'Fr_II_eTreg', 'CM_CD8', 'Th2', 'Tfh', 'Th17',\n",
    "       'Mem_CD4', 'Th1']\n",
    "\n",
    "# normalize up and down genes\n",
    "\n",
    "def standard_scale(x):\n",
    "    return (x-x.min())/x.max()\n",
    "\n",
    "#sns.heatmap(u, cmap='inferno')\n",
    "\n",
    "\n",
    "u = u.apply(standard_scale, axis=1)\n",
    "d = d.apply(standard_scale, axis=1)\n",
    "\n",
    "u = u.loc[:, ordered_cols]\n",
    "d = d.loc[:, ordered_cols]\n",
    "\n",
    "\n",
    "# collect all my cis-eqtls\n",
    "\n",
    "lookup_cis(x)\n",
    "lookup_tokyo(x)\n",
    "eqtl_catalogue_LDblock_query_type_restricted_multitype(x)\n",
    "\n",
    "# now all of these are in the Results object\n",
    "\n",
    "gwas = ldlink.ldtrait(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6b102c8a-23ed-4bf0-80e7-29cc81bfa4d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Variant_ID</th>\n",
       "      <th>Backward_slope</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>rs1354034</td>\n",
       "      <td>0.165671</td>\n",
       "      <td>Plasmablast</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene_name Variant_ID Backward_slope    cell_type\n",
       "0   ARHGEF3  rs1354034       0.165671  Plasmablast"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tokyo = x.results.tokyo_eqtl().drop_duplicates()\n",
    "tokyo = tokyo[['Gene_name','Variant_ID', 'Backward_slope', 'cell_type']]\n",
    "tokyo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "df85f5f9-192e-4e2b-bae4-e5a4f56b2a43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <td>ENSG00000163947</td>\n",
       "      <td>ENSG00000163947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>ARHGEF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant</th>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromosome</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>position</th>\n",
       "      <td>56815721</td>\n",
       "      <td>56815721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ref</th>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alt</th>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cs_id</th>\n",
       "      <td>ILMN_1781010_L1</td>\n",
       "      <td>ILMN_1781010_L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cs_index</th>\n",
       "      <td>L1</td>\n",
       "      <td>L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>finemapped_region</th>\n",
       "      <td>chr3:56079329-58079329</td>\n",
       "      <td>chr3:56079329-58079329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pip</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z</th>\n",
       "      <td>-10.5217458469795</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cs_min_r2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cs_avg_r2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cs_size</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>posterior_mean</th>\n",
       "      <td>-0.876546369440084</td>\n",
       "      <td>-0.876546369440084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>posterior_sd</th>\n",
       "      <td>0.0610111260567196</td>\n",
       "      <td>0.0610111260567196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cs_log10bf</th>\n",
       "      <td>39.766475249425</td>\n",
       "      <td>39.766475249425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>int_chrom</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study</th>\n",
       "      <td>CEDAR</td>\n",
       "      <td>CEDAR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study_type</th>\n",
       "      <td>ge</td>\n",
       "      <td>ge</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_type</th>\n",
       "      <td>platelet</td>\n",
       "      <td>platelet</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         0                       0\n",
       "molecular_trait_id         ENSG00000163947         ENSG00000163947\n",
       "gene_symbol                        ARHGEF3                 ARHGEF3\n",
       "variant                  chr3_56815721_T_C       chr3_56815721_T_C\n",
       "chromosome                               3                       3\n",
       "position                          56815721                56815721\n",
       "ref                                      T                       T\n",
       "alt                                      C                       C\n",
       "cs_id                      ILMN_1781010_L1         ILMN_1781010_L1\n",
       "cs_index                                L1                      L1\n",
       "finemapped_region   chr3:56079329-58079329  chr3:56079329-58079329\n",
       "pip                                    1.0                     1.0\n",
       "z                        -10.5217458469795                        \n",
       "cs_min_r2                              1.0                     1.0\n",
       "cs_avg_r2                              1.0                     1.0\n",
       "cs_size                                  1                       1\n",
       "posterior_mean          -0.876546369440084      -0.876546369440084\n",
       "posterior_sd            0.0610111260567196      0.0610111260567196\n",
       "cs_log10bf                 39.766475249425         39.766475249425\n",
       "int_chrom                                3                       3\n",
       "study                                CEDAR                   CEDAR\n",
       "study_type                              ge                      ge\n",
       "cell_type                         platelet                platelet"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat = x.results.eqtl_cat().drop_duplicates()\n",
    "cat.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "91b2b484-b5b3-420f-8d78-114701425c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencis = x.results.eqtlgen_cis_df.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "eeb80b80-5622-4f31-ae30-299f50f59e97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Variant_ID</th>\n",
       "      <th>Backward_slope</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>rs1354034</td>\n",
       "      <td>0.165671</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene_name Variant_ID Backward_slope\n",
       "0   ARHGEF3  rs1354034       0.165671"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2061695e-fa54-4237-820f-7d0b69624d06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the above cell does all the trans-eQTL processing and looks up all the cis-eQTL data\n",
    "\n",
    "# now to summarize this into a single figure-like thing\n",
    "\n",
    "# I need to show 4 tables (eqtlgen, eqtl_cat, immunexUT, GWAS catalog)\n",
    "\n",
    "# this tables need to be filtered down, I need to learn the syntax for showing tables on matplotlib\n",
    "\n",
    "# also the display syntax for showing seaborn clustermaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "12ded5d5-4ff0-4bd7-b184-a2daebac0a93",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'ALT'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-28-c96d2e3b59f8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0meqtl_cat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meqtl_cat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0;32mif\u001b[0m \u001b[0meqtl_cat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mALT\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEA\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32melif\u001b[0m \u001b[0meqtl_cat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mALT\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mOA\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.8/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5463\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5464\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5465\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5466\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5467\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'ALT'"
     ]
    }
   ],
   "source": [
    "# filter tables down\n",
    "\n",
    "eqtl_cat = x.results.eqtl_cat()\n",
    "\n",
    "if eqtl_cat.alt==x.EA:\n",
    "    pass\n",
    "elif eqtl_cat.alt==x.OA:\n",
    "    eqtl_cat.z = -1 * eqtl_cat.z\n",
    "    \n",
    "eqtl_cat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "752da1a8-b9f3-4ba1-b16b-49f6d2a8d8b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'T'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.EA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9bae68bc-8894-4dbe-b765-e2e2d89d6c9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>variant</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>cs_index</th>\n",
       "      <th>finemapped_region</th>\n",
       "      <th>...</th>\n",
       "      <th>cs_min_r2</th>\n",
       "      <th>cs_avg_r2</th>\n",
       "      <th>cs_size</th>\n",
       "      <th>posterior_mean</th>\n",
       "      <th>posterior_sd</th>\n",
       "      <th>cs_log10bf</th>\n",
       "      <th>int_chrom</th>\n",
       "      <th>study</th>\n",
       "      <th>study_type</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000163947</td>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>chr3_56815721_T_C</td>\n",
       "      <td>3</td>\n",
       "      <td>56815721</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ILMN_1781010_L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>chr3:56079329-58079329</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.876546369440084</td>\n",
       "      <td>0.0610111260567196</td>\n",
       "      <td>39.766475249425</td>\n",
       "      <td>3</td>\n",
       "      <td>CEDAR</td>\n",
       "      <td>ge</td>\n",
       "      <td>platelet</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  molecular_trait_id gene_symbol            variant chromosome  position ref  \\\n",
       "0    ENSG00000163947     ARHGEF3  chr3_56815721_T_C          3  56815721   T   \n",
       "\n",
       "  alt            cs_id cs_index       finemapped_region  ... cs_min_r2  \\\n",
       "0   C  ILMN_1781010_L1       L1  chr3:56079329-58079329  ...       1.0   \n",
       "\n",
       "  cs_avg_r2 cs_size      posterior_mean        posterior_sd       cs_log10bf  \\\n",
       "0       1.0       1  -0.876546369440084  0.0610111260567196  39.766475249425   \n",
       "\n",
       "  int_chrom  study study_type cell_type  \n",
       "0         3  CEDAR         ge  platelet  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.results.eqtl_cat_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9f763ec4-1df3-4409-b32d-cf5cb5a1396b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>CHR</th>\n",
       "      <th>TSS_position</th>\n",
       "      <th>Number_of_variants_cis</th>\n",
       "      <th>Variant_ID</th>\n",
       "      <th>OA</th>\n",
       "      <th>EA</th>\n",
       "      <th>Variant_CHR</th>\n",
       "      <th>Variant_position_start</th>\n",
       "      <th>Variant_position_end</th>\n",
       "      <th>Rank_of_association</th>\n",
       "      <th>Forward_nominal_P</th>\n",
       "      <th>Forward_slope</th>\n",
       "      <th>Backward_P</th>\n",
       "      <th>Backward_slope</th>\n",
       "      <th>int_chrom</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000163947.11</td>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>57079329</td>\n",
       "      <td>5997</td>\n",
       "      <td>rs1354034</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr3</td>\n",
       "      <td>56815721</td>\n",
       "      <td>56815721</td>\n",
       "      <td>0</td>\n",
       "      <td>1.75008e-05</td>\n",
       "      <td></td>\n",
       "      <td>1.75008e-05</td>\n",
       "      <td>0.165671</td>\n",
       "      <td>3</td>\n",
       "      <td>Plasmablast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000163947.11</td>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>57079329</td>\n",
       "      <td>5997</td>\n",
       "      <td>rs1354034</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr3</td>\n",
       "      <td>56815721</td>\n",
       "      <td>56815721</td>\n",
       "      <td>0</td>\n",
       "      <td>1.75008e-05</td>\n",
       "      <td></td>\n",
       "      <td>1.75008e-05</td>\n",
       "      <td>0.165671</td>\n",
       "      <td>3</td>\n",
       "      <td>Plasmablast</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene_id Gene_name   CHR TSS_position Number_of_variants_cis  \\\n",
       "0  ENSG00000163947.11   ARHGEF3  chr3     57079329                   5997   \n",
       "0  ENSG00000163947.11   ARHGEF3  chr3     57079329                   5997   \n",
       "\n",
       "  Variant_ID OA EA Variant_CHR Variant_position_start Variant_position_end  \\\n",
       "0  rs1354034  T  C        chr3               56815721             56815721   \n",
       "0  rs1354034  T  C        chr3               56815721             56815721   \n",
       "\n",
       "  Rank_of_association Forward_nominal_P Forward_slope   Backward_P  \\\n",
       "0                   0       1.75008e-05                1.75008e-05   \n",
       "0                   0       1.75008e-05                1.75008e-05   \n",
       "\n",
       "  Backward_slope int_chrom    cell_type  \n",
       "0       0.165671         3  Plasmablast  \n",
       "0       0.165671         3  Plasmablast  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.results.tokyo_eqtl_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4722e88c-5bd4-4a93-a6cb-f348f6dd3fd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pvalue</th>\n",
       "      <th>SNP</th>\n",
       "      <th>SNPChr</th>\n",
       "      <th>SNPPos</th>\n",
       "      <th>AssessedAllele</th>\n",
       "      <th>OtherAllele</th>\n",
       "      <th>Zscore</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>GeneChr</th>\n",
       "      <th>GenePos</th>\n",
       "      <th>NrCohorts</th>\n",
       "      <th>NrSamples</th>\n",
       "      <th>FDR</th>\n",
       "      <th>BonferroniP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.3396e-07</td>\n",
       "      <td>rs1354034</td>\n",
       "      <td>3</td>\n",
       "      <td>56815721</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>5.1032</td>\n",
       "      <td>ENSG00000163947</td>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>3</td>\n",
       "      <td>56937401</td>\n",
       "      <td>37</td>\n",
       "      <td>31684</td>\n",
       "      <td>0.00102408142999</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.3396e-07</td>\n",
       "      <td>rs1354034</td>\n",
       "      <td>3</td>\n",
       "      <td>56815721</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>5.1032</td>\n",
       "      <td>ENSG00000163947</td>\n",
       "      <td>ARHGEF3</td>\n",
       "      <td>3</td>\n",
       "      <td>56937401</td>\n",
       "      <td>37</td>\n",
       "      <td>31684</td>\n",
       "      <td>0.00102408142999</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Pvalue        SNP SNPChr    SNPPos AssessedAllele OtherAllele  Zscore  \\\n",
       "0  3.3396e-07  rs1354034      3  56815721              T           C  5.1032   \n",
       "0  3.3396e-07  rs1354034      3  56815721              T           C  5.1032   \n",
       "\n",
       "              Gene GeneSymbol GeneChr   GenePos NrCohorts NrSamples  \\\n",
       "0  ENSG00000163947    ARHGEF3       3  56937401        37     31684   \n",
       "0  ENSG00000163947    ARHGEF3       3  56937401        37     31684   \n",
       "\n",
       "                FDR BonferroniP  \n",
       "0  0.00102408142999         1.0  \n",
       "0  0.00102408142999         1.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.results.eqtlgen_cis()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e77e91-8d9a-44bb-8d71-33eca965c7ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23157a75-6248-4bd3-ae35-bf33a670e1b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2b17da6-bbbe-41ce-896b-e65383e8165a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b85b5b1c-8ff0-4d4e-9161-a79a9f086e83",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2d68c94-a152-402e-875d-6229a2ab4951",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7a32a675-3452-4eb6-bcb4-34cf704f5793",
   "metadata": {},
   "source": [
    "def my_plotter(ax, data1, data2, param_dict):\n",
    "    \"\"\"\n",
    "    A helper function to make a graph\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    ax : Axes\n",
    "        The axes to draw to\n",
    "\n",
    "    data1 : array\n",
    "       The x data\n",
    "\n",
    "    data2 : array\n",
    "       The y data\n",
    "\n",
    "    param_dict : dict\n",
    "       Dictionary of keyword arguments to pass to ax.plot\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    out : list\n",
    "        list of artists added\n",
    "    \"\"\"\n",
    "    out = ax.plot(data1, data2, **param_dict)\n",
    "    \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387a00a3-1612-4bed-a3da-22af37c5b96c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
