{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8abb5379-36dc-4135-8c47-c33dfd1be1a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from helpers.getpaths import *\n",
    "from helpers.ldlink import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "29f5c43d-6201-4239-b962-92a68dc5c647",
   "metadata": {},
   "outputs": [],
   "source": [
    "from helpers.ldlink import *\n",
    "\n",
    "class Variant():\n",
    "    def __init__(self, rsid:str, chrom:int, pos:int, EA:str, OA:str):\n",
    "        self.rsid = rsid \n",
    "        self.chrom = chrom\n",
    "        self.pos = pos\n",
    "        self.EA = EA\n",
    "        self.OA = OA\n",
    "    \n",
    "    def get_rsid(self):\n",
    "        return self.rsid\n",
    "    \n",
    "    def get_fullpos(self):\n",
    "        return self.chrom, self.pos\n",
    "    \n",
    "    def get_pos(self):\n",
    "        return self.pos\n",
    "    \n",
    "    def get_chrom(self):\n",
    "        return self.chrom\n",
    "    \n",
    "    def get_EA(self):\n",
    "        return self.EA\n",
    "    \n",
    "    def get_OA(self):\n",
    "        return self.OA\n",
    "    \n",
    "    def map_alleles(self, x, EA, OA):\n",
    "        d = {}\n",
    "        a1 = x.split(\",\")[0].split(\"=\")\n",
    "        a2 = x.split(\",\")[1].split(\"=\")\n",
    "        if a1[0]==EA:\n",
    "            d[EA]=a1[1]\n",
    "            d[OA]=a2[1]\n",
    "        elif a1[0]==OA:\n",
    "            d[OA] = a1[1]\n",
    "            d[EA] = a2[1]\n",
    "        return d\n",
    "    \n",
    "    def set_LDblock(self):\n",
    "        df = ldproxy(self.rsid)\n",
    "        df['EA'] = df.Correlated_Alleles.apply(lambda x:self.map_alleles(x,EA,OA)[EA])\n",
    "        df['OA'] = df.Correlated_Alleles.apply(lambda x:self.map_alleles(x,EA,OA)[OA])\n",
    "        df['chrom'] = df.Coord.apply(lambda x:int(x.split(\":\")[0][-1]))\n",
    "        df['hg38_pos'] = df.Coord.apply(lambda x:int(x.split(\":\")[1]))\n",
    "        df = df[['RS_Number', 'chrom', 'hg38_pos', 'EA','OA', 'R2', 'MAF']]\n",
    "        self.LDblock = df\n",
    "        \n",
    "rsid = 'rs10103048'\n",
    "\n",
    "chrom = 8\n",
    "pos = 130602281\n",
    "EA = 'A'\n",
    "OA = 'C'\n",
    "\n",
    "snp = Variant(rsid, chrom, pos, EA,OA)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "5c36431c-8603-4a71-8387-48f21cdbf321",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a41442bd-58eb-4c5d-957a-5c37428b787a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'eqtl_cat': '/media/cbio3/data/eQTL_DB/',\n",
       " 'dbsnp': '/media/cbio3/data/dbSNP/',\n",
       " 'eqtl_tokyo': '/media/cbio3/projects/Zain_2021/ImmuNEXT_Japan_Cell2021/eQTL_summarystats/',\n",
       " 'ge_tokyo': '/media/cbio3/projects/Zain_2021/ImmunexUT_GE/E-GEAD-397.processed/tpm/',\n",
       " 'eqtlgen_cis': '/media/cbio3/projects/Zain_2021/eQTLgen/data/2019-12-11-cis-eQTLsFDR0.05-ProbeLevel-CohortInfoRemoved-BonferroniAdded_sorted.txt ',\n",
       " 'eqtlgen_trans': '/media/cbio3/projects/Zain_2021/eQTLgen/data/2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt ',\n",
       " 'tokyo_alleles': '/media/cbio3/projects/Zain_2021/ImmuNEXT_Japan_Cell2021/metadata/full_allele_defs.txt',\n",
       " 'tokyo_ge': '/media/cbio3/projects/Zain_2021/ImmunexUT_GE/E-GEAD-397.processed/tpm/',\n",
       " 'ensembl': '/media/cbio3/data/ensembl_biomart/gene_aliases.txt',\n",
       " 'ucsc': '/media/cbio3/data/ensembl_biomart/hg38_genes_ucsc.txt'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paths = get_paths(\"/media/\")\n",
    "\n",
    "paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bf116dee-35ee-4a64-a837-4330bc63c6b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid = 'rs10103048'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fcbe8295-6b89-4916-a780-065e927abc2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now each of these resources needs different methods for access\n",
    "\n",
    "import os\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "\n",
    "# this is waiting some kind of tabix indexing of the file \n",
    "\n",
    "def get_eqtlgen_cis(snp_list, paths=paths):\n",
    "    # need to call get_paths first to get eqtlgen_cis filepath\n",
    "    \n",
    "    # function that simply filters eQTLgen-cis to list of rsids\n",
    "    \n",
    "    x = pd.read_csv(paths['eqtlgen_cis'], sep='\\t')\n",
    "    \n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c5daafba-8a21-49d0-93c0-e5435f2a6f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tokyo_cis(snp_list,paths=paths):\n",
    "    \n",
    "    # loops through all the sumstats and pulls out all matches to query SNPs\n",
    "    \n",
    "    x = os.listdir(paths['eqtl_tokyo'])\n",
    "    x = [paths['eqtl_tokyo'] + i for i in x]\n",
    "    \n",
    "    df_list = []\n",
    "    \n",
    "    for file in x:\n",
    "        df = pd.read_csv(file, sep='\\t')\n",
    "        \n",
    "        df = df[df.Variant_ID.isin(snp_list)]\n",
    "        \n",
    "        cell = file.split(\"/\")[-1].split(\"_cond\")[0]\n",
    "        \n",
    "        df['cell'] = cell\n",
    "        \n",
    "        df_list.append(df)\n",
    "    return pd.concat(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a91218d4-5ba9-485d-be66-72069c8516b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>molecular_trait_id</th>\n",
       "      <th>variant</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>cs_index</th>\n",
       "      <th>finemapped_region</th>\n",
       "      <th>pip</th>\n",
       "      <th>z</th>\n",
       "      <th>cs_min_r2</th>\n",
       "      <th>cs_avg_r2</th>\n",
       "      <th>cs_size</th>\n",
       "      <th>posterior_mean</th>\n",
       "      <th>posterior_sd</th>\n",
       "      <th>cs_log10bf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172</td>\n",
       "      <td>chr1_1000079_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>1000079</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172_L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>chr1:142-2000142</td>\n",
       "      <td>0.037619</td>\n",
       "      <td>-5.673441</td>\n",
       "      <td>0.546196</td>\n",
       "      <td>0.790143</td>\n",
       "      <td>19</td>\n",
       "      <td>-0.028499</td>\n",
       "      <td>0.146938</td>\n",
       "      <td>4.039558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172</td>\n",
       "      <td>chr1_1000112_G_T</td>\n",
       "      <td>1</td>\n",
       "      <td>1000112</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172_L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>chr1:142-2000142</td>\n",
       "      <td>0.037339</td>\n",
       "      <td>-5.671394</td>\n",
       "      <td>0.546196</td>\n",
       "      <td>0.790143</td>\n",
       "      <td>19</td>\n",
       "      <td>-0.028300</td>\n",
       "      <td>0.146485</td>\n",
       "      <td>4.039558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172</td>\n",
       "      <td>chr1_1000291_C_G</td>\n",
       "      <td>1</td>\n",
       "      <td>1000291</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172_L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>chr1:142-2000142</td>\n",
       "      <td>0.084505</td>\n",
       "      <td>5.849436</td>\n",
       "      <td>0.546196</td>\n",
       "      <td>0.790143</td>\n",
       "      <td>19</td>\n",
       "      <td>0.075840</td>\n",
       "      <td>0.254031</td>\n",
       "      <td>4.039558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172</td>\n",
       "      <td>chr1_1000453_C_G</td>\n",
       "      <td>1</td>\n",
       "      <td>1000453</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172_L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>chr1:142-2000142</td>\n",
       "      <td>0.037265</td>\n",
       "      <td>-5.670987</td>\n",
       "      <td>0.546196</td>\n",
       "      <td>0.790143</td>\n",
       "      <td>19</td>\n",
       "      <td>-0.028206</td>\n",
       "      <td>0.146154</td>\n",
       "      <td>4.039558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172</td>\n",
       "      <td>chr1_1000814_A_G</td>\n",
       "      <td>1</td>\n",
       "      <td>1000814</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000188290.10_1_1000112_1000172_L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>chr1:142-2000142</td>\n",
       "      <td>0.039019</td>\n",
       "      <td>-5.683077</td>\n",
       "      <td>0.546196</td>\n",
       "      <td>0.790143</td>\n",
       "      <td>19</td>\n",
       "      <td>-0.029735</td>\n",
       "      <td>0.150412</td>\n",
       "      <td>4.039558</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     molecular_trait_id           variant chromosome  \\\n",
       "0  ENSG00000188290.10_1_1000112_1000172  chr1_1000079_A_G          1   \n",
       "1  ENSG00000188290.10_1_1000112_1000172  chr1_1000112_G_T          1   \n",
       "2  ENSG00000188290.10_1_1000112_1000172  chr1_1000291_C_G          1   \n",
       "3  ENSG00000188290.10_1_1000112_1000172  chr1_1000453_C_G          1   \n",
       "4  ENSG00000188290.10_1_1000112_1000172  chr1_1000814_A_G          1   \n",
       "\n",
       "   position ref alt                                    cs_id cs_index  \\\n",
       "0   1000079   A   G  ENSG00000188290.10_1_1000112_1000172_L1       L1   \n",
       "1   1000112   G   T  ENSG00000188290.10_1_1000112_1000172_L1       L1   \n",
       "2   1000291   C   G  ENSG00000188290.10_1_1000112_1000172_L1       L1   \n",
       "3   1000453   C   G  ENSG00000188290.10_1_1000112_1000172_L1       L1   \n",
       "4   1000814   A   G  ENSG00000188290.10_1_1000112_1000172_L1       L1   \n",
       "\n",
       "  finemapped_region       pip         z  cs_min_r2  cs_avg_r2  cs_size  \\\n",
       "0  chr1:142-2000142  0.037619 -5.673441   0.546196   0.790143       19   \n",
       "1  chr1:142-2000142  0.037339 -5.671394   0.546196   0.790143       19   \n",
       "2  chr1:142-2000142  0.084505  5.849436   0.546196   0.790143       19   \n",
       "3  chr1:142-2000142  0.037265 -5.670987   0.546196   0.790143       19   \n",
       "4  chr1:142-2000142  0.039019 -5.683077   0.546196   0.790143       19   \n",
       "\n",
       "   posterior_mean  posterior_sd  cs_log10bf  \n",
       "0       -0.028499      0.146938    4.039558  \n",
       "1       -0.028300      0.146485    4.039558  \n",
       "2        0.075840      0.254031    4.039558  \n",
       "3       -0.028206      0.146154    4.039558  \n",
       "4       -0.029735      0.150412    4.039558  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# eqtl Catalogue - need to remap all my credible set summary files to proper gene and variant names using the ref files provided \n",
    "# downloading them now \n",
    "\n",
    "def get_eqtlcat_credset(snp_list, paths=paths):\n",
    "    pass\n",
    "\n",
    "all_files = os.listdir(paths['eqtl_cat'] + 'credible_sets')\n",
    "\n",
    "meta = [i for i in os.listdir(paths['eqtl_cat']) if i.split(\".\")[-1]==\"csv\"]\n",
    "meta = paths['eqtl_cat'] + meta[0]\n",
    "meta = pd.read_csv(meta).Study.tolist()\n",
    "\n",
    "files = []\n",
    "\n",
    "for i in all_files:\n",
    "    for m in meta:\n",
    "        if m in i:\n",
    "            files.append(paths['eqtl_cat']+'credible_sets/' + i)\n",
    "            \n",
    "x = pd.read_csv(files[0], sep='\\t')\n",
    "x.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9beb0b3b-4080-49df-8cf0-796715032aaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>NW100766</th>\n",
       "      <th>NW101334</th>\n",
       "      <th>NW101409</th>\n",
       "      <th>NW102537</th>\n",
       "      <th>NW103721</th>\n",
       "      <th>NW103984</th>\n",
       "      <th>NW107509</th>\n",
       "      <th>NW108551</th>\n",
       "      <th>NW109068</th>\n",
       "      <th>...</th>\n",
       "      <th>NW189242</th>\n",
       "      <th>NW191139</th>\n",
       "      <th>NW191977</th>\n",
       "      <th>NW192678</th>\n",
       "      <th>NW192956</th>\n",
       "      <th>NW194176</th>\n",
       "      <th>NW197021</th>\n",
       "      <th>NW199501</th>\n",
       "      <th>NW199834</th>\n",
       "      <th>NW199907</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>71.026893</td>\n",
       "      <td>85.991016</td>\n",
       "      <td>109.480961</td>\n",
       "      <td>69.982619</td>\n",
       "      <td>79.777627</td>\n",
       "      <td>66.310846</td>\n",
       "      <td>73.788173</td>\n",
       "      <td>67.735969</td>\n",
       "      <td>66.084273</td>\n",
       "      <td>...</td>\n",
       "      <td>100.272661</td>\n",
       "      <td>67.969278</td>\n",
       "      <td>66.091513</td>\n",
       "      <td>80.695705</td>\n",
       "      <td>68.617867</td>\n",
       "      <td>104.447285</td>\n",
       "      <td>74.442871</td>\n",
       "      <td>71.514384</td>\n",
       "      <td>60.559690</td>\n",
       "      <td>64.104525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>10.497253</td>\n",
       "      <td>8.561123</td>\n",
       "      <td>9.890112</td>\n",
       "      <td>13.879714</td>\n",
       "      <td>12.345047</td>\n",
       "      <td>9.779591</td>\n",
       "      <td>14.080985</td>\n",
       "      <td>11.952519</td>\n",
       "      <td>14.041186</td>\n",
       "      <td>...</td>\n",
       "      <td>10.968706</td>\n",
       "      <td>12.112270</td>\n",
       "      <td>14.370151</td>\n",
       "      <td>8.736413</td>\n",
       "      <td>13.935659</td>\n",
       "      <td>8.617643</td>\n",
       "      <td>11.142311</td>\n",
       "      <td>15.275693</td>\n",
       "      <td>8.200183</td>\n",
       "      <td>11.187171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>9.231251</td>\n",
       "      <td>6.386137</td>\n",
       "      <td>6.472582</td>\n",
       "      <td>9.386585</td>\n",
       "      <td>8.818284</td>\n",
       "      <td>6.084325</td>\n",
       "      <td>9.681862</td>\n",
       "      <td>9.659316</td>\n",
       "      <td>10.725835</td>\n",
       "      <td>...</td>\n",
       "      <td>9.563534</td>\n",
       "      <td>7.300959</td>\n",
       "      <td>10.581185</td>\n",
       "      <td>6.783008</td>\n",
       "      <td>10.602472</td>\n",
       "      <td>8.989581</td>\n",
       "      <td>9.064241</td>\n",
       "      <td>8.132379</td>\n",
       "      <td>6.853539</td>\n",
       "      <td>10.224853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FGR</td>\n",
       "      <td>1169.062819</td>\n",
       "      <td>1517.581161</td>\n",
       "      <td>1303.774565</td>\n",
       "      <td>1093.915984</td>\n",
       "      <td>1291.692031</td>\n",
       "      <td>1328.939290</td>\n",
       "      <td>1402.950905</td>\n",
       "      <td>1040.949908</td>\n",
       "      <td>1354.458503</td>\n",
       "      <td>...</td>\n",
       "      <td>1294.194750</td>\n",
       "      <td>1110.638627</td>\n",
       "      <td>1281.460222</td>\n",
       "      <td>1436.458546</td>\n",
       "      <td>1324.341811</td>\n",
       "      <td>1182.974232</td>\n",
       "      <td>1446.580401</td>\n",
       "      <td>1423.035437</td>\n",
       "      <td>1131.913962</td>\n",
       "      <td>1281.657631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CFH</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.137726</td>\n",
       "      <td>0.417020</td>\n",
       "      <td>0.178547</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.140630</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.138282</td>\n",
       "      <td>0.053769</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.293720</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.145480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53339</th>\n",
       "      <td>AL139254.3</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53340</th>\n",
       "      <td>AL591163.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53341</th>\n",
       "      <td>AL589702.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53342</th>\n",
       "      <td>AL034417.4</td>\n",
       "      <td>0.413797</td>\n",
       "      <td>0.130119</td>\n",
       "      <td>0.074015</td>\n",
       "      <td>0.821797</td>\n",
       "      <td>1.066418</td>\n",
       "      <td>0.426147</td>\n",
       "      <td>0.940324</td>\n",
       "      <td>0.755207</td>\n",
       "      <td>0.500826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.443912</td>\n",
       "      <td>0.584411</td>\n",
       "      <td>0.790514</td>\n",
       "      <td>0.633443</td>\n",
       "      <td>0.742598</td>\n",
       "      <td>0.705829</td>\n",
       "      <td>0.238073</td>\n",
       "      <td>0.525778</td>\n",
       "      <td>0.924488</td>\n",
       "      <td>0.347224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53343</th>\n",
       "      <td>AL513220.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.116298</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53344 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene_name     NW100766     NW101334     NW101409     NW102537  \\\n",
       "0            DPM1    71.026893    85.991016   109.480961    69.982619   \n",
       "1           SCYL3    10.497253     8.561123     9.890112    13.879714   \n",
       "2        C1orf112     9.231251     6.386137     6.472582     9.386585   \n",
       "3             FGR  1169.062819  1517.581161  1303.774565  1093.915984   \n",
       "4             CFH     0.000000     0.000000     0.000000     0.137726   \n",
       "...           ...          ...          ...          ...          ...   \n",
       "53339  AL139254.3     0.000000     0.000000     0.000000     0.000000   \n",
       "53340  AL591163.1     0.000000     0.000000     0.000000     0.000000   \n",
       "53341  AL589702.1     0.000000     0.000000     0.000000     0.000000   \n",
       "53342  AL034417.4     0.413797     0.130119     0.074015     0.821797   \n",
       "53343  AL513220.1     0.000000     0.000000     0.000000     0.000000   \n",
       "\n",
       "          NW103721     NW103984     NW107509     NW108551     NW109068  ...  \\\n",
       "0        79.777627    66.310846    73.788173    67.735969    66.084273  ...   \n",
       "1        12.345047     9.779591    14.080985    11.952519    14.041186  ...   \n",
       "2         8.818284     6.084325     9.681862     9.659316    10.725835  ...   \n",
       "3      1291.692031  1328.939290  1402.950905  1040.949908  1354.458503  ...   \n",
       "4         0.417020     0.178547     0.000000     0.140630     0.000000  ...   \n",
       "...            ...          ...          ...          ...          ...  ...   \n",
       "53339     0.000000     0.000000     0.000000     0.000000     0.000000  ...   \n",
       "53340     0.000000     0.000000     0.000000     0.000000     0.000000  ...   \n",
       "53341     0.000000     0.000000     0.000000     0.000000     0.000000  ...   \n",
       "53342     1.066418     0.426147     0.940324     0.755207     0.500826  ...   \n",
       "53343     0.000000     0.000000     0.000000     0.000000     0.000000  ...   \n",
       "\n",
       "          NW189242     NW191139     NW191977     NW192678     NW192956  \\\n",
       "0       100.272661    67.969278    66.091513    80.695705    68.617867   \n",
       "1        10.968706    12.112270    14.370151     8.736413    13.935659   \n",
       "2         9.563534     7.300959    10.581185     6.783008    10.602472   \n",
       "3      1294.194750  1110.638627  1281.460222  1436.458546  1324.341811   \n",
       "4         0.000000     0.000000     0.000000     0.000000     0.138282   \n",
       "...            ...          ...          ...          ...          ...   \n",
       "53339     0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "53340     0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "53341     0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "53342     0.443912     0.584411     0.790514     0.633443     0.742598   \n",
       "53343     0.000000     0.000000     0.000000     0.000000     0.000000   \n",
       "\n",
       "          NW194176     NW197021     NW199501     NW199834     NW199907  \n",
       "0       104.447285    74.442871    71.514384    60.559690    64.104525  \n",
       "1         8.617643    11.142311    15.275693     8.200183    11.187171  \n",
       "2         8.989581     9.064241     8.132379     6.853539    10.224853  \n",
       "3      1182.974232  1446.580401  1423.035437  1131.913962  1281.657631  \n",
       "4         0.053769     0.000000     0.293720     0.000000     0.145480  \n",
       "...            ...          ...          ...          ...          ...  \n",
       "53339     0.000000     0.000000     0.000000     0.000000     0.000000  \n",
       "53340     0.000000     0.000000     0.000000     0.000000     0.000000  \n",
       "53341     0.000000     0.000000     0.000000     0.000000     0.000000  \n",
       "53342     0.705829     0.238073     0.525778     0.924488     0.347224  \n",
       "53343     0.000000     0.000000     0.000000     0.116298     0.000000  \n",
       "\n",
       "[53344 rows x 79 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gene expression lookups in ImmunexUT data\n",
    "\n",
    "# multiple donors for every cell types\n",
    "\n",
    "# extract list of numbers for each cell type. Can use this to construct boxplots, means, etc.\n",
    "\n",
    "\n",
    "g = os.listdir(paths['ge_tokyo'])\n",
    "\n",
    "g = [paths['ge_tokyo'] + i for i in g]\n",
    "\n",
    "pd.read_csv(g[0], sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "24daeb47-e16a-4bed-b7a0-1e687287e7c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Thu Aug 18 17:33:47 2022 WARNING ID 'AL513220.1' not found. Please double-check spelling/arguments and try again.\n"
     ]
    }
   ],
   "source": [
    "import gget\n",
    "\n",
    "gget.info([\"AL513220.1\"], \"homo_sapiens\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6edddb1f-dcf5-48ec-896b-93503506d574",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remapping gene names to ensembl IDs for Gex lookups\n",
    "\n",
    "def gene2ensembl(gene, paths=paths):\n",
    "    \n",
    "    # take common gene name and return ensembl ID\n",
    "    # checks all synonyms as well\n",
    "    \n",
    "    # load ensembl mappings into dataframe \n",
    "    x = pd.read_csv(paths['ensembl'], sep='\\t')\n",
    "    \n",
    "    x.columns = ['ensembl', 'version', 'name', 'synonym', 'HGNC']\n",
    "    \n",
    "    # if the gene name matches the HGNC name then there will be a string match\n",
    "    # on the name column\n",
    "    \n",
    "    if len(x[x.name==gene])>0:\n",
    "        x = set(x[x.name==gene].ensembl.tolist())\n",
    "        x = list(x)\n",
    "        assert len(x)==1\n",
    "        return x[0]\n",
    "    \n",
    "    # otherwise we go to the synonyms column\n",
    "    \n",
    "    elif len(x[x.synonym==gene])>0:\n",
    "        w = x[x.synonym==gene]\n",
    "        \n",
    "        assert len(w)==1\n",
    "        return w.name.values[0]\n",
    "    \n",
    "def ensembl2gene(ensembl, paths=paths):\n",
    "    x = pd.read_csv(paths['ensembl'], sep='\\t')\n",
    "    x.columns = ['ensembl', 'version', 'name', 'synonym', 'HGNC']\n",
    "    x = x[x.ensembl==ensembl]\n",
    "    \n",
    "    x = set(x.name.tolist())\n",
    "    x = list(x)\n",
    "    assert len(x)==1\n",
    "    \n",
    "    return x[0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "544c0571-4d5a-4498-b68c-02c67d6efc39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Previous symbol</th>\n",
       "      <th>Alias symbol</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "      <th>NCBI gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>NCRNA00181</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>503538.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>A1BGAS</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>503538.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>A1BG-AS</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>503538.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>29974.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72674</th>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain contain...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KIAA0399</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>23140.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72675</th>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain contain...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZZZ4</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>23140.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72676</th>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain contain...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLJ10821</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>23140.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72677</th>\n",
       "      <td>HGNC:24523</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>zinc finger ZZ-type containing 3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DKFZP564I052</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>26009.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72678</th>\n",
       "      <td>HGNC:24523</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>zinc finger ZZ-type containing 3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATAC1</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>26009.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72679 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          HGNC ID Approved symbol  \\\n",
       "0          HGNC:5            A1BG   \n",
       "1      HGNC:37133        A1BG-AS1   \n",
       "2      HGNC:37133        A1BG-AS1   \n",
       "3      HGNC:37133        A1BG-AS1   \n",
       "4      HGNC:24086            A1CF   \n",
       "...           ...             ...   \n",
       "72674  HGNC:29027           ZZEF1   \n",
       "72675  HGNC:29027           ZZEF1   \n",
       "72676  HGNC:29027           ZZEF1   \n",
       "72677  HGNC:24523            ZZZ3   \n",
       "72678  HGNC:24523            ZZZ3   \n",
       "\n",
       "                                           Approved name Previous symbol  \\\n",
       "0                                 alpha-1-B glycoprotein             NaN   \n",
       "1                                   A1BG antisense RNA 1      NCRNA00181   \n",
       "2                                   A1BG antisense RNA 1          A1BGAS   \n",
       "3                                   A1BG antisense RNA 1         A1BG-AS   \n",
       "4                         APOBEC1 complementation factor             NaN   \n",
       "...                                                  ...             ...   \n",
       "72674  zinc finger ZZ-type and EF-hand domain contain...             NaN   \n",
       "72675  zinc finger ZZ-type and EF-hand domain contain...             NaN   \n",
       "72676  zinc finger ZZ-type and EF-hand domain contain...             NaN   \n",
       "72677                   zinc finger ZZ-type containing 3             NaN   \n",
       "72678                   zinc finger ZZ-type containing 3             NaN   \n",
       "\n",
       "       Alias symbol  Ensembl gene ID  NCBI gene ID  \n",
       "0               NaN  ENSG00000121410           1.0  \n",
       "1          FLJ23569  ENSG00000268895      503538.0  \n",
       "2          FLJ23569  ENSG00000268895      503538.0  \n",
       "3          FLJ23569  ENSG00000268895      503538.0  \n",
       "4               ACF  ENSG00000148584       29974.0  \n",
       "...             ...              ...           ...  \n",
       "72674      KIAA0399  ENSG00000074755       23140.0  \n",
       "72675          ZZZ4  ENSG00000074755       23140.0  \n",
       "72676      FLJ10821  ENSG00000074755       23140.0  \n",
       "72677  DKFZP564I052  ENSG00000036549       26009.0  \n",
       "72678         ATAC1  ENSG00000036549       26009.0  \n",
       "\n",
       "[72679 rows x 7 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y = pd.read_csv(paths['ucsc'], sep ='\\t')\n",
    "y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "101319fa-785a-4641-97e9-5b9a511d320b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Previous symbol</th>\n",
       "      <th>Alias symbol</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "      <th>NCBI gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [HGNC ID, Approved symbol, Approved name, Previous symbol, Alias symbol, Ensembl gene ID, NCBI gene ID]\n",
       "Index: []"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y[y['Ensembl gene ID']=='ENSG00000284747']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a184c59e-f4f1-4c08-8bd9-4101ad414643",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Previous symbol</th>\n",
       "      <th>Alias symbol</th>\n",
       "      <th>Ensembl gene ID</th>\n",
       "      <th>NCBI gene ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>29974.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASP</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>29974.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF64</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>29974.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF65</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>29974.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>APOBEC1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>29974.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      HGNC ID Approved symbol                   Approved name Previous symbol  \\\n",
       "4  HGNC:24086            A1CF  APOBEC1 complementation factor             NaN   \n",
       "5  HGNC:24086            A1CF  APOBEC1 complementation factor             NaN   \n",
       "6  HGNC:24086            A1CF  APOBEC1 complementation factor             NaN   \n",
       "7  HGNC:24086            A1CF  APOBEC1 complementation factor             NaN   \n",
       "8  HGNC:24086            A1CF  APOBEC1 complementation factor             NaN   \n",
       "\n",
       "  Alias symbol  Ensembl gene ID  NCBI gene ID  \n",
       "4          ACF  ENSG00000148584       29974.0  \n",
       "5          ASP  ENSG00000148584       29974.0  \n",
       "6        ACF64  ENSG00000148584       29974.0  \n",
       "7        ACF65  ENSG00000148584       29974.0  \n",
       "8    APOBEC1CF  ENSG00000148584       29974.0  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y[y['Ensembl gene ID']=='ENSG00000148584']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e33ba966-232f-4d0e-84b2-4a070e251b33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#chrom', 'chromStart', 'chromEnd', 'name', 'score', 'strand',\n",
       "       'thickStart', 'thickEnd', 'reserved', 'blockCount', 'blockSizes',\n",
       "       'chromStarts', 'name2', 'cdsStartStat', 'cdsEndStat', 'exonFrames',\n",
       "       'type', 'geneName', 'geneName2', 'geneType', 'transcriptClass',\n",
       "       'source', 'transcriptType', 'tag', 'level', 'tier'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "14cb9d49-a310-44fb-afe5-5e5546f63cb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Thu Aug 18 16:28:08 2022 WARNING ID 'AL513220.1' not found. Please double-check spelling/arguments and try again.\n"
     ]
    }
   ],
   "source": [
    "gget.info(\"AL513220.1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "656993ed-6b6d-40f3-aa7d-9250faf8b380",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
